---
title: "R Notebook"
---

Using ArchR for some steps due to memory issues in Seurat.

```{r Define directories}
HH6_dir <- "/data/Austin/10XscATAC/HH6_10xscATAC_Lane1/outs/"
HH8_dir <- "/data/Austin/10XscATAC/HH8_10xscATAC_Lane1_Lane2/outs/"
HH10_dir <- "/data/Austin/10XscATAC/HH10_10xscATAC_Lane1_Lane2/outs/"
HH12_dir <- "/data/Austin/10XscATAC/HH12_10xscATAC_Lane1_Lane2/outs/"
HH14_dir <- "/data/Austin/10XscATAC/HH14_10xscATAC_Lane1_Lane2/outs/"
HH16_dir <- "/data/Austin/10XscATAC/HH16_10xscATAC_Lane1/outs/"
HH18_dir <- "/data/Austin/10XscATAC/HH18_10xscATAC_Lane1/outs/"
libraries <- c("HH6","HH8","HH10","HH12","HH14","HH16","HH18")
```

```{r message=FALSE, warning=FALSE}
library(ArchR)
library(BSgenome.Ggallus.ENSEMBL.galGal6)
library(org.Gg.eg.db)
library(GenomicFeatures)
library(dplyr)
library(qdapTools)
library(plotly)
library(ggplot2)
library(parallel)
addArchRThreads(threads = 62) 
set.seed(1)

# Make a blacklist for AADN05001427.1
arch_gg6 <- createGenomeAnnotation(genome = BSgenome.Ggallus.ENSEMBL.galGal6, filter = F)
# Manually remove noncontigous chromosomes.
arch_gg6$chromSizes <- arch_gg6$chromSizes[1:35]
TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 103)
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, annoStyle = "ENSEMBL")
```

Now build the Project. (One time only using v1.0.1)
```{r}
setwd("/data/Austin/10XscATAC/ArchR_data_final/")

all_fragments <- list.files(path = "/data/Austin/10XscATAC/ArchR_data_final", full.names = T, pattern = "fragments.tsv.gz$")
names(all_fragments) <- c("HH10","HH12","HH14","HH16","HH18","HH6","HH8")

# Ensure names are matched with fragments!!!
all_fragments

# TODO there appears to be a bug with fragments that extend past the bounds of the custom genome. See https://github.com/GreenleafLab/ArchR/issues/1145
ArrowFiles <- createArrowFiles(
  inputFiles = all_fragments,
  sampleNames = names(all_fragments),
  geneAnnotation = arch_gg6_genes,
  genomeAnnotation = arch_gg6,
  addTileMat = T,
  minTSS = 3,
  minFrags = 1000,
  nChunk = 5,
  excludeChr = c("MT"),
  force = T, threads = 62)

projNC <- ArchRProject(
  ArrowFiles = ArrowFiles, 
  outputDirectory = "/data/Austin/10XscATAC/ArchR_data_final/",
  copyArrows = T,
  geneAnnotation = arch_gg6_genes,
  genomeAnnotation = arch_gg6)
```

Let's look at doublet detection and filter these out.
```{r}
# Due to an error with the 1.0.1 version, we will load the information from addDoubletScores run with version 1.0.2.

# projNC_1.0.2 <- readRDS(file = "/data/Austin/10XscATAC/ArchR_data_save/Save-ArchR-Project.rds")
# 
# doublet_results <- data.frame(cellNames = projNC_1.0.2$cellNames,
#   DoubletScore = projNC_1.0.2$DoubletScore,
#   DoubletEnrichment = projNC_1.0.2$DoubletEnrichment)
# 
# 
# projNC$DoubletScore <- lookup(terms = projNC$cellNames,
#                               doublet_results[,1:2])
# 
# projNC$DoubletEnrichment <- lookup(terms = projNC$cellNames,
#                               doublet_results[,1:3])

# Let's try doublet detection with release 1.0.3
projNC <- addDoubletScores(
  projNC,
  useMatrix = "TileMatrix",
  k = 10,
  nTrials = 10,
  force = T,
  threads = 62)

projNC <- addIterativeLSI(
  projNC, useMatrix = "TileMatrix",
  iterations = 5,
  varFeatures = 35000,
  clusterParams = list(resolution = c(2),
                       sampleCells = 10000,
                       maxClusters = 8, n.start = 10),
  UMAPParams = list(n_neighbors = 30, min_dist = 0.2,
                    metric = "cosine", verbose =FALSE,
                    fast_sgd = TRUE, repulsion_strength = 1.25),
  force = T)

projNC <- addUMAP(
    ArchRProj = projNC, 
    reducedDims = "IterativeLSI", 
    name = "UMAP", 
    nNeighbors = 32, 
    minDist = 0.25, 
    metric = "cosine",
    force = T,
    n_components = 2,
    saveModel = F,
    repulsion_strength = 1.25)

# Add in the peak information/matrix
HH6_peaks <- readRDS(file = paste0(HH6_dir, "macs3/HH6_combined_peaks.RDS"))
HH8_peaks <- readRDS(file = paste0(HH8_dir, "macs3/HH8_combined_peaks.RDS"))
HH10_peaks <- readRDS(file = paste0(HH10_dir, "macs3/HH10_combined_peaks.RDS"))
HH12_peaks <- readRDS(file = paste0(HH12_dir, "macs3/HH12_combined_peaks.RDS"))
HH14_peaks <- readRDS(file = paste0(HH14_dir, "macs3/HH14_combined_peaks.RDS"))
HH16_peaks <- readRDS(file = paste0(HH16_dir, "macs3/HH16_combined_peaks.RDS"))
HH18_peaks <- readRDS(file = paste0(HH18_dir, "macs3/HH18_combined_peaks.RDS"))

combined.peaks <- reduce(c(HH6_peaks, HH8_peaks, HH10_peaks,HH12_peaks, HH14_peaks, HH16_peaks, HH18_peaks), drop.empty.ranges = T)

summary(width(combined.peaks))

projNC <- addPeakSet(projNC, peakSet = combined.peaks, force = T)

projNC <- addPeakMatrix(projNC, binarize = T, threads = 32, force = T)

projNC$batch <- ifelse(projNC$Sample %in% c("HH8","HH10","HH14","HH16"),
                       "batch1",
                       ifelse(projNC$Sample %in% c("HH6","HH18"),
                              "batch2","batch3"))

#   SAVE POINT HERE, USED FOR PRE-FILTERING PLOTS.
saveArchRProject(projNC, outputDirectory = "/data/Austin/10XscATAC/ArchR_Prefiltered")
```

Great, we have created a basic ArchR project, but we haven't performed any filtering yet. So, We will start from this Prefiltered object and apply cell QC filtering, and re-run the pipeline.

```{r}
projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_Prefiltered")

# Using the built-in filterDoublets command.
projNC_tmp <- filterDoublets(projNC)

doublet_cells <- projNC$cellNames[!projNC$cellNames %in% projNC_tmp$cellNames]

write.csv(doublet_cells, file = "doublet_cells.csv")
rm(projNC_tmp)
projNC <- filterDoublets(projNC)

# Now let's filter by the metrics determined from other QC plots.

projNC <- projNC[projNC$TSSEnrichment >= 3.5]
projNC <- projNC[projNC$FRIP >= 0.4]
projNC <- projNC[log10(projNC$nFrags) >= 3.3]

saveArchRProject(projNC,
                 outputDirectory = "/data/Austin/10XscATAC/ArchR_data_save_2",
                 threads = 1)
```

After filtering, we have to repeat everything.

```{r}
projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_save_2")
```

What about motif enrichment? First, let's get differential peaks.

```{r}
# Run chromVAR! 
projNC <- addMotifAnnotations(
  projNC, motifSet = "JASPAR2020",
  name = "Motif", collection = "CORE", force = T, tax_group = "vertebrates")

projNC <- addBgdPeaks(projNC, method = "ArchR", force = T)

# 1.0.2 issues with multithreading/writing hd5 files?
# Error when trying to overwrite an existing MotifMatrix. Make sure arrow files are empty beforehand.

projNC <- addDeviationsMatrix(
  ArchRProj = projNC, binarize = T,
  peakAnnotation = "Motif",
  force = T, out = "z")

```

```{r}
projNC <- addIterativeLSI(
  projNC, useMatrix = "PeakMatrix",
  iterations = 3,
  varFeatures = 30000,
  clusterParams = list(resolution = c(2),
                       sampleCells = 10000,
                       maxClusters = 10, n.start = 10),
  UMAPParams = list(n_neighbors = 40, min_dist = 0.22,
                    metric = "cosine", verbose =FALSE,
                    fast_sgd = TRUE, repulsion_strength = 1.25),
  force = T)

projNC <- addUMAP(
    ArchRProj = projNC, 
    reducedDims = "IterativeLSI",
    dimsToUse = c(1:30),
    name = "UMAP", 
    nNeighbors = 30, 
    minDist = 0.22, 
    spread = 0.8,
    metric = "cosine",
    force = T,
    n_components = 2,
    saveModel = F,
    repulsion_strength = 1.2)

projNC <- addUMAP(
    ArchRProj = projNC, 
    reducedDims = "IterativeLSI",
    dimsToUse = c(1:30),
    name = "UMAP_3d", 
    nNeighbors = 30, 
    minDist = 0.22, 
    spread = 0.8,
    metric = "cosine",
    force = T,
    n_components = 3,
    saveModel = F,
    repulsion_strength = 1.2)

projNC <- addClusters(
    input = projNC,
    reducedDims = "IterativeLSI",
    method = "Seurat",
    name = "Clusters",
    dimsToUse = c(1:30),
    resolution = 0.38,
    knnAssign = 30,
    nOutlier = 20,
    n.iter = 20,
    force = T)

# 2D
p1 <- plotEmbedding(ArchRProj = projNC, colorBy = "cellColData", name = "Sample", embedding = "UMAP", size = 1, baseSize = 14)
p1

# 3D
df_3d <- getEmbedding(projNC, embedding = "UMAP_3d", returnDF = TRUE)
colnames(df_3d) <- c("UMAP_1","UMAP_2","UMAP_3")
df_3d$Sample <- projNC$Sample
df_3d$SeuratClusters <- projNC$Clusters
df_3d$Pseudotime <- projNC$hour_adj_pseudotime

p1_3d <- plot_ly(df_3d, x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, color = ~Sample, size = I(14)) %>% add_markers()
p2_3d <- plot_ly(df_3d, x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, color = ~Pseudotime, size = I(10)) %>% add_markers()


p2 <- plotEmbedding(ArchRProj = projNC, colorBy = "cellColData", name = "Clusters", embedding = "UMAP", size = 1, baseSize = 14)
p2

cowplot::plot_grid(p1 + p2)
```
The 3D UMAP projections using Iterative LSI are great, similar to Harmony integration via batch in Seurat. I can't quite find the rotation that might be best yet. That's fine, it is more a visual thing we can tweak later.

Next, let's get into Gene Scores with the new ArchR algorithm.

```{r}
projNC <- addGeneScoreMatrix(
  projNC, extendUpstream = c(5000, 1e+05),
  extendDownstream = c(5000,1e+05),
  geneUpstream = 2000, force = T)

projNC <- addImputeWeights(projNC, sampleCells = 10000, nRep = 3, k = 20)


markersGS <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "Clusters",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    testMethod = "wilcoxon")

markerList <- getMarkers(markersGS, cutOff = "FDR <= 0.01 & Log2FC >= 1.25")

markerDF <- as.data.frame(markerList)

top_genes <- markerDF %>% dplyr::group_by(group_name) %>% dplyr::top_n(10, wt = Log2FC)

markerGenes  <- c(
    "SOX10", 
    "TFAP2A", 
    "PAX7",
    "PRRX2",
    "PBX4")

heatmapGS <- plotMarkerHeatmap(
  seMarker = markersGS, 
  cutOff = "FDR <= 0.01 & Log2FC >= 1.25", 
  labelMarkers = markerGenes,
  transpose = T)

# Heatmap function is currently broken.
# ComplexHeatmap::draw_heatmap_list(heatmapGS, heatmap_legend_side = "bot", annotation_legend_side = "bot")

# plotPDF(heatmapGS, name = "GeneScores-Marker-Heatmap", width = 8, height = 6, ArchRProj = projNC, addDOC = FALSE)

p3 <- plotEmbedding(
    ArchRProj = projNC, 
    colorBy = "GeneScoreMatrix", 
    name = top_genes$name, rastr = F, sampleCells = NULL, plotAs = "points",
    embedding = "UMAP",
    quantCut = c(0.05, 0.9),
    imputeWeights = getImputeWeights(projNC))

# Let's look at local genome browser

p4 <- plotBrowserTrack(
    ArchRProj = projNC, 
    groupBy = "Clusters", 
    geneSymbol = "PRRX2", 
    upstream = 50000,
    downstream = 50000)

grid::grid.newpage()
grid::grid.draw(p4$PRRX2)
```

```{r}
# Get the best motifs per cluster.
Cluster_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    bias = c("log10(nFrags)"),
    testMethod = "wilcoxon",
    useSeqnames = "z")

Motif_list <- getMarkers(Cluster_Motifs, cutOff = "FDR <= 0.01")
Motif_df <- as.data.frame(Motif_list)

Motif_df$TF <- stringr::str_split(Motif_df$name, pattern = "_", simplify = T)[,1]
write_csv(Motif_df, file = "Figure_Exports/TF_Markers.csv")

top_Motifs <- Motif_df %>% dplyr::group_by(group_name) %>% dplyr::top_n(3, wt = Log2FC)
```

Let's set up a couple of pseudotime trajectory arms for the major clusters we find. We could define these as having an overrepresentation of HH18 cells, but it's easier to just look and go by the clusters. 

```{r}
# C4 is HH6/root
trajectory1 <- c("C4", "C9") # Sensory neurons
trajectory2 <- c("C4", "C2") # Sensory Ganglia
trajectory3 <- c("C4", "C1") # Melanocytes
trajectory4 <- c("C4", "C3","C8","C10") # Mesenchyme
trajectory5 <- c("C4", "C3","C5") # Hindbrain Neurons
trajectory6 <- c("C4", "C7") # Midbrain Neurons
trajectory7 <- c("C4", "C6") # Somatosensory Neurons

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C9END", 
    groupBy = "Clusters",
    trajectory = trajectory1, 
    embedding = "UMAP", 
    preFilterQuantile = 1, 
    postFilterQuantile = 1,
    force = TRUE)

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C2END", 
    groupBy = "Clusters",
    trajectory = trajectory2, 
    embedding = "UMAP", 
    preFilterQuantile = 1, 
    postFilterQuantile = 1,
    force = TRUE)

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C1END", 
    groupBy = "Clusters",
    trajectory = trajectory3, 
    embedding = "UMAP", 
    preFilterQuantile = 1, 
    postFilterQuantile = 1,
    force = TRUE)

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C10END", 
    groupBy = "Clusters",
    trajectory = trajectory4,
    embedding = "UMAP", 
    preFilterQuantile = 1, 
    postFilterQuantile = 1,
    force = TRUE)

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C5END", 
    groupBy = "Clusters",
    trajectory = trajectory5, 
    embedding = "UMAP",
    preFilterQuantile = 1, 
    postFilterQuantile = 1, 
    force = TRUE)

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C7END", 
    groupBy = "Clusters",
    trajectory = trajectory6, 
    embedding = "UMAP", 
    preFilterQuantile = 1, 
    postFilterQuantile = 1,
    force = TRUE)

projNC <- addTrajectory(
    ArchRProj = projNC, 
    name = "C6END", 
    groupBy = "Clusters",
    trajectory = trajectory7, 
    embedding = "UMAP", 
    preFilterQuantile = 1, 
    postFilterQuantile = 1,
    force = TRUE)

# Fix an issue with retrieving an embedding with >2 dimensions.
# plotTrajectoryCustom <- edit(plotTrajectory)

# It might be more accurate to 
# Generate a score based on the 'average' age of the cells in these clusters.
# What is the hours between each stage?
# HH6, HH8, HH10, HH12, HH14, HH16, HH18
# 24 28 36 46 52 54 60
hh_to_hours <- data.frame(timepoint = c("HH6","HH8","HH10","HH12",
                                        "HH14","HH16","HH18"),
                          hours = c(24,28,36,46,52,54,60))

# For every cell of that stage, multiply by the hours of the stage and then average them out for an average cluster TRUE age.

C9_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C9",]
C2_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C2",]
C1_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C1",]
C10_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C10",]
C7_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C7",]
C5_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C5",]
C6_cells <- projNC@cellColData[projNC@cellColData$Clusters == "C6",]

avg_hours <- function(x){
  df <- data.frame(timepoint = x$Sample@values,
           cells = x$Sample@lengths)
  df$hours <- lookup(df$timepoint, hh_to_hours)
  return(sum(df[,2] * df[,3]) / sum(df[,2]))
}

C9_weighted_age <- avg_hours(C9_cells)
C2_weighted_age <- avg_hours(C2_cells)
C1_weighted_age <- avg_hours(C1_cells)
C10_weighted_age <- avg_hours(C10_cells)
C7_weighted_age <- avg_hours(C7_cells)
C5_weighted_age <- avg_hours(C5_cells)
C6_weighted_age <- avg_hours(C6_cells)

C9END_Weighted <- as.numeric(projNC@cellColData[,"C9END"] * C9_weighted_age)
C2END_Weighted <- as.numeric(projNC@cellColData[,"C2END"] * C2_weighted_age)
C1END_Weighted <- as.numeric(projNC@cellColData[,"C1END"] * C1_weighted_age)
C10END_Weighted <- as.numeric(projNC@cellColData[,"C10END"] * C10_weighted_age)
C7END_Weighted <- as.numeric(projNC@cellColData[,"C7END"] * C7_weighted_age)
C5END_Weighted <- as.numeric(projNC@cellColData[,"C5END"] * C5_weighted_age)
C6END_Weighted <- as.numeric(projNC@cellColData[,"C6END"] * C6_weighted_age)

# Are there instances of NA despite including every cluster?
test <- data.frame(
  C9END_Weighted,
  C2END_Weighted,
  C1END_Weighted,
  C10END_Weighted,
  C7END_Weighted,
  C5END_Weighted,
  C6END_Weighted
)

summary(is.na(test[,1:7]))

test$avg <- Matrix::rowMeans(test, na.rm = T)

test$rescaled <-  round(scales::rescale(test$avg, to = c(1, 100)))

projNC$hour_adj_pseudotime <- test$rescaled

p <- plotTrajectory(projNC, trajectory = "hour_adj_pseudotime",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP")

p[[1]]

# WOW! This is really nice.
```

## CoAccessibility
Let's run coaccessability with ArchR!

```{r}
# What are the best motifs over time?
projNC <- addCoAccessibility(
    ArchRProj = projNC,
    reducedDims = "IterativeLSI")

cA <- getCoAccessibility(
    ArchRProj = projNC,
    corCutOff = 0.5,
    resolution = 1,
    returnLoops = FALSE)

saveArchRProject(ArchRProj = projNC, outputDirectory = "/data/Austin/10XscATAC/ArchR_data_final_thesis", load = TRUE)
```

## Differential Markers
```{r}
# Marker Peaks
Peak_Markers <- getMarkerFeatures(projNC, groupBy = "Clusters", useMatrix = "PeakMatrix")
Peak_Markers.df <- getMarkers(Peak_Markers, cutOff = "FDR <= 0.05 & Log2FC >= 0.5")
Peak_Markers.df <- as(unlist(Peak_Markers.df), "data.frame")
Peak_Markers.df$Cluster <- stringr::str_split(rownames(Peak_Markers.df), pattern = fixed("."), simplify = T)[,1]
write_csv(Peak_Markers.df, file = "Figure_Exports/Peak_Markers.csv")

# Export each cluster's enriched peaks as a bed file
for (cl in unique(Peak_Markers.df$Cluster)){
  bed <- Peak_Markers.df[Peak_Markers.df$Cluster == cl,c("seqnames","start","end")]
  export.bed(bed, con = paste0("Figure_Exports/Peak_Marker_Beds/",cl,".bed"))
}

# Marker Gene Scores
Gene_Markers <- getMarkerFeatures(projNC, groupBy = "Clusters", useMatrix = "GeneScoreMatrix")
Gene_Markers.df <- getMarkers(Gene_Markers, cutOff = "FDR <= 0.05 & Log2FC >= 0.5")
Gene_Markers.df <- as(unlist(Gene_Markers.df), "data.frame")
Gene_Markers.df$Cluster <- stringr::str_split(rownames(Gene_Markers.df), pattern = fixed("."), simplify = T)[,1]
write_csv(Gene_Markers.df, file = "Figure_Exports/Gene_Markers.csv")
```

