---
title: "R Notebook"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
    fig_width: 8
    fig_height: 6
editor_options:
  chunk_output_type: inline
---

Exploration of building GRN's using scATAC data.

So basically, the idea is to combine the positive/negative regulator prediction from motif presence and peak activity over time to build a network of transcription factor regulation. This way, we can assign an average strength and direction to the motif presence within any given peak. 

There are a few ways we wish to filter this. 1. We want to look at highly specific peaks. 2. We want to consider only those highly specific peaks that also are linked to specific genes (via Genescore). These may be too restrictive initially, but we will see! 

So it turns out the Genescore isn't related to the peak coaccessability at all. It's based on tile-matrix.
```{r Loading libraries and data, message=FALSE, warning=FALSE}
library(ArchR)
library(knitr)
library(TFBSTools)
library(ComplexHeatmap)
library(ggplot2)
library(ggrepel)
library(Seurat)
library(Signac)
library(BSgenome.Ggallus.ENSEMBL.galGal6)
library(GenomicFeatures)
library(org.Gg.eg.db)
library(qdapTools)
library(stringr)
library(readr)
library(qdapTools)

addArchRThreads(threads = 62) 
set.seed(1)
projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_final_thesis/")

projNC@projectMetadata$outputDirectory <- "./Figure_Exports"
projNC <- addImputeWeights(projNC, sampleCells = 10000, nRep = 3, k = 20)
cluster_to_identity <- data.frame(cluster = paste0("C",1:10),
                                  identity = c("C1_Melanocytes","C2_Sensory_Glia","C3_Int_Mesenchyme",
                                               "C4_Multipotent_NC","C5_Hindbrain_Neurons","C6_Somatosensory_Neurons",
                                               "C7_Midbrain_Neurons","C8_Early_Mesenchyme","C9_Sensory_Neurons",
                                               "C10_Late_Mesenchyme"))
projNC$Cell_Identity <- lookup(projNC$Clusters, cluster_to_identity)

TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 103)
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, annoStyle = "ENSEMBL")
ensembl2symbol <- as.data.frame(mcols(arch_gg6_genes$genes))
ensembl2symbol$chr <- as.character(seqnames(arch_gg6_genes$genes))
```

```{r}
# Need to name the peaks in the peakset in order to identify.
ps <- projNC@peakSet
ps$name <- paste0(seqnames(ps),"_peak",ps$idx)
projNC@peakSet <- ps
projNC <- addPeakMatrix(ArchRProj = projNC, force = TRUE, threads = 62)
```


Let's explore the CoAccessibility object. Specifically, we are looking for a way to link peaks to genes. I think the easiest way is just to check if either of the peaks overlap the gene's promoter region (-5kb to +1kb).

```{r coAccess by lineage}
# Make a GRange object with promoters.
galGal6_prom <- trim(promoters(genes(TxDb_galGal6), upstream = 5000, downstream = 1000))
galGal6_prom$gene_symbol <- lookup(terms = galGal6_prom$gene_id, key.match = ensembl2symbol)
peakset <- getPeakSet(projNC)
All_Peak_Regions <- as.data.frame(peakset)
All_Peak_Regions$idx <- as.character(All_Peak_Regions$idx)
All_Peak_Regions$PeakID <- paste0(All_Peak_Regions$seqnames, "_", All_Peak_Regions$start, "_", All_Peak_Regions$end)

getLineageCoAccessibility <- function(ArchRProj = projNC,
                                      cells = "",
                                      corCutOff = 0.3,
                                      resolution = 1){
  # Subset ArchR project to just that lineage
  lineage_Proj <- addCoAccessibility(ArchRProj = ArchRProj,
                                     reducedDims = "IterativeLSI",
                                     cellsToUse = cells, 
                                     maxDist = 5e5,
                                     threads = 62,
                                     seed = "05010501")
  # Get coAccessible elements
  cA <- as.data.frame(getCoAccessibility(
      ArchRProj = lineage_Proj,
      corCutOff = 0.3,
      resolution = 1,
      returnLoops = F))
  # Annotate with PeakID
  b <- All_Peak_Regions[,c("idx","PeakID")]
  b <- b[cA$queryHits,]
  cA$Peak1 <- b$PeakID
  c <- All_Peak_Regions[,c("idx","PeakID")]
  c <- c[cA$subjectHits,]
  cA$Peak2 <- c$PeakID
  
  # Which coAccessible elements have a peak in a promoter?
  Peak1_coords <- str_split(cA$Peak1, pattern = "_", n = 3, simplify = T)
  Peak1_ovl <- makeGRangesFromDataFrame(df = data.frame(chr = Peak1_coords[,1],
                                                      start = Peak1_coords[,2],
                                                      end = Peak1_coords[,3]))
  hits <- findOverlaps(Peak1_ovl, galGal6_prom)

  Peak1_ovl$gene_id <- ifelse(Peak1_ovl %in% Peak1_ovl[hits@from], galGal6_prom[hits@to]$gene_id, NA)


  Peak2_coords <- str_split(cA$Peak2, pattern = "_", n = 3, simplify = T)
  Peak2_ovl <- makeGRangesFromDataFrame(df = data.frame(chr = Peak2_coords[,1],
                                                      start = Peak2_coords[,2],
                                                      end = Peak2_coords[,3]))
  hits2 <- findOverlaps(Peak2_ovl, galGal6_prom)

  Peak2_ovl$gene_id <- ifelse(Peak2_ovl %in% Peak2_ovl[hits@from], galGal6_prom[hits@to]$gene_id, NA)

  cA$Overlap1 <- Peak1_ovl$gene_id
  cA$Overlap2 <- Peak2_ovl$gene_id

  # Annotate as Distal-Distal, Distal-Prom, or Prom-Prom
  cA$Overlap_Type <- ifelse(is.na(cA$Overlap1) & is.na(cA$Overlap2), "Distal-Distal",
                          ifelse(!is.na(cA$Overlap1) & !is.na(cA$Overlap2), "Promoter-Promoter",
                                 "Distal-Promoter"))

  cA$Symbol1 <-  qdapTools::lookup(cA$Overlap1, ensembl2symbol)
  cA$Symbol2 <-  qdapTools::lookup(cA$Overlap2, ensembl2symbol)
  
  return(cA)
}

CorrelatedToGene <- function(coaccess = cA, goi = "", cutoff = 0.3){
  # A function to select identify all of the peaks that are associated with a certain gene given a correlation cutoff.
  # Inputs
  # coaccess - A formatted coaccessability report with at least Peak1, Peak2, Symbol1, Symbol2, and coaccessibility measurements
  # goi - Gene of Interest, an ENSEMBL ID of a gene you wish to find peaks associated with.
  # cutoff - a numerical variable to filter corrleation values.
  
  # Find the ensembl ID of the requested overlap
  ensembl <- lookup(goi, ensembl2symbol[,c(2,1)])
  chr <- lookup(goi, ensembl2symbol[,c(2,3)])
  # Subset the coaccess dataset for Symbol overlaps
  coaccess1 <- coaccess[goi %in% coaccess$Symbol1,]
   # Subset the coaccess dataset for only the chromosome that the gene is on.
  coaccess1 <- coaccess1[str_split(coaccess1$Overlap1, pattern = "_", simplify = T)[,1] == chr & 
                         str_split(coaccess1$Overlap2, pattern = "_", simplify = T)[,1] == chr,]
  # Filter out to correlation cutoff
  coaccess1 <- coaccess1[max(coaccess1[3:9], na.rm = T) >= cutoff,]
  print(length(coaccess1$Symbol1and2))
  coaccess2 <- coaccess[goi %in% coaccess$Symbol2,]
  
  coaccess2 <- coaccess2[str_split(coaccess2$Overlap1, pattern = "_", simplify = T)[,1] == chr & 
                         str_split(coaccess2$Overlap2, pattern = "_", simplify = T)[,1] == chr,]
  
  # Filter out to correlation cutoff
  coaccess2 <- coaccess2[max(coaccess2[3:9], na.rm = T) >= cutoff,]
  print(length(coaccess2$Symbol1and2))
  
  #return(coaccess_comb)
}


```

```{r Generate CoAccessibility per lineage}
# C2 END (sensory Glia CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C2END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C2END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C2END_cA_filt <- C2END_cA[C2END_cA$correlation > 0.75,]
write_csv(C2END_cA_filt, file = "Figure_Exports/C2END_cA_500Kb_0.75.csv")

# C9 END (sensory neuron CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C9END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C9END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C9END_cA_filt <- C9END_cA[C9END_cA$correlation > 0.75,]
write_csv(C9END_cA_filt, file = "Figure_Exports/C9END_cA_500Kb_0.75.csv")

# C1 END (Melanocyte CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C1END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C1END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C1END_cA_filt <- C1END_cA[C1END_cA$correlation > 0.75,]
write_csv(C1END_cA_filt, file = "Figure_Exports/C1END_cA_500Kb_0.75.csv")

# C10 END (Mesenchyme CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C10END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C10END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C10END_cA_filt <- C10END_cA[C10END_cA$correlation > 0.75,]
write_csv(C10END_cA_filt, file = "Figure_Exports/C10END_cA_500Kb_0.75.csv")

# C5 END (Hindbrain CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C5END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C5END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C5END_cA_filt <- C5END_cA[C5END_cA$correlation > 0.75,]
write_csv(C5END_cA_filt, file = "Figure_Exports/C5END_cA_500Kb_0.75.csv")

# C6 END (Somatosensory CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C6END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C6END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C6END_cA_filt <- C6END_cA[C6END_cA$correlation > 0.75,]
write_csv(C6END_cA_filt, file = "Figure_Exports/C6END_cA_500Kb_0.75.csv")

# C7 END (Somatosensory CoAccessibility)
lineage_cells <- rownames(projNC@cellColData)[projNC$C7END >0]
lineage_cells <- lineage_cells[!is.na(lineage_cells)]
C7END_cA <- getLineageCoAccessibility(ArchRProj = projNC, cells = lineage_cells)
C7END_cA_filt <- C7END_cA[C7END_cA$correlation > 0.75,]
write_csv(C7END_cA_filt, file = "Figure_Exports/C7END_cA_500Kb_0.75.csv")

# Summary stats of lineage-level CoAccessibility
C2END_cA$Peak1_and_Peak2 <- paste0(C2END_cA$Peak1,"-",C2END_cA$Peak2)
C9END_cA$Peak1_and_Peak2 <- paste0(C9END_cA$Peak1,"-",C9END_cA$Peak2)
C1END_cA$Peak1_and_Peak2 <- paste0(C1END_cA$Peak1,"-",C1END_cA$Peak2)
C10END_cA$Peak1_and_Peak2 <- paste0(C10END_cA$Peak1,"-",C10END_cA$Peak2)
C5END_cA$Peak1_and_Peak2 <- paste0(C5END_cA$Peak1,"-",C5END_cA$Peak2)
C6END_cA$Peak1_and_Peak2 <- paste0(C6END_cA$Peak1,"-",C6END_cA$Peak2)
C7END_cA$Peak1_and_Peak2 <- paste0(C7END_cA$Peak1,"-",C7END_cA$Peak2)

C2END_cA$Symbol1_and_Symbol2 <- paste0(C2END_cA$Symbol1,"-",C2END_cA$Symbol2)
C9END_cA$Symbol1_and_Symbol2 <- paste0(C9END_cA$Symbol1,"-",C9END_cA$Symbol2)
C1END_cA$Symbol1_and_Symbol2 <- paste0(C1END_cA$Symbol1,"-",C1END_cA$Symbol2)
C10END_cA$Symbol1_and_Symbol2 <- paste0(C10END_cA$Symbol1,"-",C10END_cA$Symbol2)
C5END_cA$Symbol1_and_Symbol2 <- paste0(C5END_cA$Symbol1,"-",C5END_cA$Symbol2)
C6END_cA$Symbol1_and_Symbol2 <- paste0(C6END_cA$Symbol1,"-",C6END_cA$Symbol2)
C7END_cA$Symbol1_and_Symbol2 <- paste0(C7END_cA$Symbol1,"-",C7END_cA$Symbol2)

# Adding annotations
All_Peak_Combinations_anno = data.frame(
  All_Peak_Combinations = c(
  C2END_cA$Peak1_and_Peak2,C9END_cA$Peak1_and_Peak2,C1END_cA$Peak1_and_Peak2,
  C10END_cA$Peak1_and_Peak2,C5END_cA$Peak1_and_Peak2,C6END_cA$Peak1_and_Peak2,
  C7END_cA$Peak1_and_Peak2),
  Symbol1and2 = c(
  C2END_cA$Symbol1_and_Symbol2,C9END_cA$Symbol1_and_Symbol2,C1END_cA$Symbol1_and_Symbol2,
  C10END_cA$Symbol1_and_Symbol2,C5END_cA$Symbol1_and_Symbol2,C6END_cA$Symbol1_and_Symbol2,
  C7END_cA$Symbol1_and_Symbol2))

head(All_Peak_Combinations_anno)

All_Peak_Combinations_anno$C2END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C2END_cA[,c("Peak1_and_Peak2","correlation")])
All_Peak_Combinations_anno$C9END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C9END_cA[,c("Peak1_and_Peak2","correlation")])
All_Peak_Combinations_anno$C1END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C1END_cA[,c("Peak1_and_Peak2","correlation")])
All_Peak_Combinations_anno$C10END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C10END_cA[,c("Peak1_and_Peak2","correlation")])
All_Peak_Combinations_anno$C5END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C5END_cA[,c("Peak1_and_Peak2","correlation")])
All_Peak_Combinations_anno$C6END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C6END_cA[,c("Peak1_and_Peak2","correlation")])
All_Peak_Combinations_anno$C7END_Correlation <- lookup(All_Peak_Combinations_anno$All_Peak_Combinations, C7END_cA[,c("Peak1_and_Peak2","correlation")])

merged_CoAccess_matrix <- as.matrix(All_Peak_Combinations_anno[,3:9])
All_Peak_Combinations_anno$Lineage_Variability <- MatrixGenerics::rowVars(merged_CoAccess_matrix, na.rm = T)

All_Peak_Combinations_anno_ordered <- All_Peak_Combinations_anno[order(All_Peak_Combinations_anno$Lineage_Variability, decreasing = T),]

write_csv(All_Peak_Combinations_anno_ordered, "Figure_Exports/CoAccessibility_Lineage_Summary.csv")
quantile(All_Peak_Combinations_anno_ordered$Lineage_Variability, prob=c(.75,.9, .95), na.rm = T)

# Top 5% variable peaks
All_Peak_Combinations_anno_filt <- All_Peak_Combinations_anno_ordered[All_Peak_Combinations_anno_ordered$Lineage_Variability > 0.03067606,]

write_csv(All_Peak_Combinations_anno_filt, "Figure_Exports/CoAccessibility_Lineage_top0.05-variable_Summary.csv")
```

## Summary of different CoAccessibility in lineages
```{r}
All_Peak_Combinations_anno_filt <- read_csv("Figure_Exports/CoAccessibility_Lineage_top0.05-variable_Summary.csv")

All_Peak_Combinations_anno_filt$Overlap1 <- str_split(All_Peak_Combinations_anno_filt$All_Peak_Combinations, pattern = "-", simplify = T)[,1]
All_Peak_Combinations_anno_filt$Overlap2 <- str_split(All_Peak_Combinations_anno_filt$All_Peak_Combinations, pattern = "-", simplify = T)[,2]

All_Peak_Combinations_anno_filt$Symbol1 <- str_split(All_Peak_Combinations_anno_filt$Symbol1and2, pattern = "-", simplify = T)[,1]
All_Peak_Combinations_anno_filt$Symbol2 <- str_split(All_Peak_Combinations_anno_filt$Symbol1and2, pattern = "-", simplify = T)[,2]

All_Peak_Combinations_anno_filt$Overlap1_chr <- str_split(All_Peak_Combinations_anno_filt$Overlap1, pattern = "_", simplify = T)[,1]
All_Peak_Combinations_anno_filt$Overlap1_start <- str_split(All_Peak_Combinations_anno_filt$Overlap1, pattern = "_", simplify = T)[,2]
All_Peak_Combinations_anno_filt$Overlap1_end <- str_split(All_Peak_Combinations_anno_filt$Overlap1, pattern = "_", simplify = T)[,3]

All_Peak_Combinations_anno_filt$Overlap2_chr <- str_split(All_Peak_Combinations_anno_filt$Overlap2, pattern = "_", simplify = T)[,1]
All_Peak_Combinations_anno_filt$Overlap2_start <- str_split(All_Peak_Combinations_anno_filt$Overlap2, pattern = "_", simplify = T)[,2]
All_Peak_Combinations_anno_filt$Overlap2_end <- str_split(All_Peak_Combinations_anno_filt$Overlap2, pattern = "_", simplify = T)[,3]


Sox10_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("SOX10"), cutoff = 0.1)
# Replace NA with 0.001
Sox10_Coaccess[is.na(Sox10_Coaccess)] <- 0.001

# standard barplots of a pair in different lineages.
Sox10_Coaccess_melted <- melt(Sox10_Coaccess, na.rm = T)
Sox10_Coaccess_melted <- Sox10_Coaccess_melted[!Sox10_Coaccess_melted$variable %in% c("Lineage_Variability",
                                                                                      "C6END_Correlation",
                                                                                      "C5END_Correlation",
                                                                                      "C1END_Correlation"),]

# Highlight chosen pairs
chosen_pairs <- c("1_8865093_8865717-1_9334105_9334884","1_8770765_8771511-1_9021962_9022767",
                  "1_8770765_8771511-1_8278449_8278522","1_8865093_8865717-1_9352463_9352972",
                  "1_8770765_8771511-1_8892470_8893355",
                  "1_8865093_8865717-1_9352463_9352972","1_8876665_8877710-1_8854403_8855264",
                  "1_8876665_8877710-1_8382605_8383539","1_8874571_8875138-1_8387736_8388342")
Sox10_Coaccess_melted <- Sox10_Coaccess_melted[Sox10_Coaccess_melted$All_Peak_Combinations %in% chosen_pairs,]

ggplot(Sox10_Coaccess_melted, aes(x = All_Peak_Combinations, y = value, group = variable, fill = variable)) +
  geom_col(position = "dodge") + ggtitle("SOX10 Locus CoAccessibility by Lineage")

ggsave(filename = "Figure_Exports/Plots/Sox10_CoAccessibility_by_Lineage_Barchart.pdf", width = 8, height = 5)
```

## Overlap of CoAccessibility and previously cloned enhancer regions

```{r}
library(GenomicRanges)
library(IRanges)
cloned_enhancers <- read_csv(file = "cloned_enhancers.csv")
cloned_enhancers.gr <- makeGRangesFromDataFrame(cloned_enhancers)
cloned_enhancers.gr$name <- cloned_enhancers$Name

Overlap1.df <- data.frame(chr = All_Peak_Combinations_anno_filt$Overlap1_chr,
             start = as.numeric(All_Peak_Combinations_anno_filt$Overlap1_start),
             end = as.numeric(All_Peak_Combinations_anno_filt$Overlap1_end))
Overlap1.df <- Overlap1.df[!is.na(Overlap1.df$chr),]

Overlap1.gr <- makeGRangesFromDataFrame(Overlap1.df)

enhancer_overlaps <- findOverlaps(query = cloned_enhancers.gr,
             subject = Overlap1.gr)

enhancer_to_CoAccess <- data.frame(cloned_location = cloned_enhancers.gr[enhancer_overlaps@from],
                                   Overlap1_location = Overlap1.gr[enhancer_overlaps@to])
enhancer_to_CoAccess$Cloned <- paste0(enhancer_to_CoAccess$cloned_location.seqnames,"_",
                                            enhancer_to_CoAccess$cloned_location.start,"_",
                                            enhancer_to_CoAccess$cloned_location.end)
enhancer_to_CoAccess$Overlap1 <- paste0(enhancer_to_CoAccess$Overlap1_location.seqnames,"_",
                                            enhancer_to_CoAccess$Overlap1_location.start,"_",
                                            enhancer_to_CoAccess$Overlap1_location.end)

enhancer_to_CoAccess <- unique(enhancer_to_CoAccess)
enhancer_to_CoAccess <- enhancer_to_CoAccess[,c("cloned_location.name","Cloned","Overlap1")]

# Annotate
enhancer_to_CoAccess <- merge(enhancer_to_CoAccess, All_Peak_Combinations_anno_filt, by = "Overlap1")
enhancer_to_CoAccess$Overlap1_Symbol <- lookup(enhancer_to_CoAccess$Overlap1_Peak, as.data.frame(All_Peak_Combinations_anno_filt[,c("Overlap1","Symbol1")]))

# Overlap 2?

Overlap2.df <- data.frame(chr = All_Peak_Combinations_anno_filt$Overlap2_chr,
             start = as.numeric(All_Peak_Combinations_anno_filt$Overlap2_start),
             end = as.numeric(All_Peak_Combinations_anno_filt$Overlap2_end))
Overlap2.df <- Overlap2.df[!is.na(Overlap2.df$chr),]
Overlap2.df <- Overlap2.df[!is.na(Overlap2.df$start),]
Overlap2.df <- Overlap2.df[!is.na(Overlap2.df$end),]

Overlap2.gr <- makeGRangesFromDataFrame(Overlap2.df)

enhancer_overlaps2 <- findOverlaps(query = cloned_enhancers.gr,
             subject = Overlap2.gr)

enhancer_to_CoAccess2 <- data.frame(cloned_location = cloned_enhancers.gr[enhancer_overlaps2@from],
                                   Overlap2_location = Overlap2.gr[enhancer_overlaps2@to])
enhancer_to_CoAccess2$Cloned <- paste0(enhancer_to_CoAccess2$cloned_location.seqnames,"_",
                                            enhancer_to_CoAccess2$cloned_location.start,"_",
                                            enhancer_to_CoAccess2$cloned_location.end)
enhancer_to_CoAccess2$Overlap2 <- paste0(enhancer_to_CoAccess2$Overlap2_location.seqnames,"_",
                                            enhancer_to_CoAccess2$Overlap2_location.start,"_",
                                            enhancer_to_CoAccess2$Overlap2_location.end)

enhancer_to_CoAccess2 <- unique(enhancer_to_CoAccess2)
enhancer_to_CoAccess2 <- enhancer_to_CoAccess2[,c("cloned_location.name","Cloned","Overlap2")]

# Annotate
enhancer_to_CoAccess2 <- merge(enhancer_to_CoAccess2, All_Peak_Combinations_anno_filt, by = "Overlap2")
enhancer_to_CoAccess2$Overlap1_Symbol <- lookup(enhancer_to_CoAccess2$Overlap1_Peak, as.data.frame(All_Peak_Combinations_anno_filt[,c("Overlap2","Symbol2")]))

# Combine both tables

combined_cloned_enhancer_anno <- merge(enhancer_to_CoAccess, enhancer_to_CoAccess2, all.x = T, all.y = T)

write_csv(combined_cloned_enhancer_anno, file = "Cloned_Enhancers_CoAcessibility_Annotated.csv")

```

```{r}
# Let's do some UMAP plots of enhancer activity
# C8 Glial B 2_42661783_42662544
# C2 E 27_5656219_5656425
# C6 Neural_1_E 1_173791114_173792531
# C7 M8b  1_147724181_147725058
# Neural_1_K 1_21759441_21760749

peak_of_interest.gr <- makeGRangesFromDataFrame(data.frame(chr = "1",
                                    start = "147724181",
                                    end = "147725058"))
ovl <- findOverlaps(peak_of_interest.gr,ps)

en_plot <- plotEmbedding(
  ArchRProj = projNC,
  colorBy = "PeakMatrix",
  name = ps[ovl@to]$name,
  embedding = "UMAP",
  quantCut = c(0.01, 0.95),
  imputeWeights = getImputeWeights(ArchRProj = projNC),
  rastr = TRUE,
  plotAs = "points"
)


plotPDF(en_plot, ArchRProj = projNC, name = "C7_Midbrain_Neurons_M8B_Enhancer_UMAP.pdf", addDOC = FALSE, width = 5, height = 5)
```

## Investigation of GRN in mesenchyme lineage.

```{r}
TF_Markers <- read_csv("Figure_Exports/TF_Markers.csv")
# Mesenchyme lineage

mes_tf_list <- c("TFAP2B_300","FLI1_133","ETS1_251",
                 "TWIST1_741","NR2F1_220","ZBTB18_192")



# Get CoAccessibile peaks in the mesenchyme lineage for each of the genes.
TFAP2B_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("TFAP2B"), cutoff = 0.1)
FLI1_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("FLI1"), cutoff = 0.1)
ETS1_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("ETS1"), cutoff = 0.1)
TWIST1_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("TWIST1"), cutoff = 0.1)
NR2F1_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("NR2F1"), cutoff = 0.1)
ZBTB18_Coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt, goi = c("ZBTB18"), cutoff = 0.1)

# Binary presence of TF motif?
Motif_in_Peak <- readRDS("/data/Austin/10XscATAC/ArchR_data_final_backup/Annotations/Motif-In-Peaks-Summary.rds")

findMotifOverlaps <- function(tf = "", all_tfs = "",CoAccess = All_Peak_Combinations_anno_filt) {
  tf_symbol <- str_split(tf, pattern = "_", simplify = T)[,1]
  
  tf_coaccess <- CorrelatedToGene(coaccess = All_Peak_Combinations_anno_filt,
                                  goi = tf_symbol,
                                  cutoff = 0.1)
  
  overlap1.df <- tf_coaccess[,c("Overlap1_chr","Overlap1_start","Overlap1_end")]
  colnames(overlap1.df) <- c("chr","start","end")
  overlap1.gr <- makeGRangesFromDataFrame(overlap1.df)
  
  # look at all factors in the lineage
  for(each in all_tfs) {
      tf.gr <- Motif_in_Peak$motifPositions[each]
      # Count the number of overlaps that the motif has in the Overlap1 gr
      overlaps <- findOverlaps(tf.gr, overlap1.gr)
      print(paste0(tf_symbol, " has ", length(overlaps[,2]), " CoAccessible regions with a motif for ",
                                              each))
  }
}

FLI1_motif_summary <- findMotifOverlaps(tf = "TWIST1_741", all_tfs = mes_tf_list)


```



```{r}
# Trying to make a custom loop plot below, not working.
prep_gr <- Sox10_Coaccess[Sox10_Coaccess$C9END_Correlation > 0.4,c("Overlap1_chr","Overlap1_start","Overlap1_end","C9END_Correlation")]

prep_gr <- prep_gr[!is.na(prep_gr$Overlap1_chr),]
colnames(prep_gr) <- c("chr","start","end","value")
prep_gr$correlation <- prep_gr$value

Sox10_Coaccess_C9END <- SimpleList(CoAccessibility = makeGRangesFromDataFrame(prep_gr,keep.extra.columns = T))

p1 <- plotBrowserTrack(
    ArchRProj = projNC, 
    groupBy = "Clusters", 
    geneSymbol = c("SOX10"), 
    upstream = 250000,
    downstream = 250000,
    loops = Sox10_Coaccess_C9END)

grid::grid.draw(p1$SOX10)


edit(plotBrowserTrack)

## Initiate writing to PDF file
pdf("./Figure_Exports/Plots/SOX10_CoAccessibility.pdf", height = 6, width = 8)
grid::grid.draw(p1$SOX10)
dev.off()



# Example loops object
ex_loops <- getCoAccessibility(projNC, corCutOff = 0.7, returnLoops = T)
```

