---
title: "NC Multiome"
output: html_notebook
---

# NC Multiome

## Setup

```{r}
library(renv)
renv::activate()
# install.packages("BSgenome.Ggallus.ENSEMBL.galGal6_1.0.3.tar.gz", repos = NULL)

suppressPackageStartupMessages({
  library(ArchR)
  library(BSgenome.Ggallus.ENSEMBL.galGal6)
  library(GenomicFeatures)
  library(RMariaDB)
  library(org.Gg.eg.db)
})

addArchRThreads(threads = 44)
set.seed(1)
```

Genome Annotation

```{r}
# Make a blacklist for AADN05001427.1
# Manually remove noncontigous chromosomes.
# TxDb_galGal6 <- makeTxDbFromGFF("/workspaces/NC_scATAC/genes.gtf.gz",
#   organism = "Gallus gallus", dataSource = "Ensembl", chrominfo=seqinfo(BSgenome.Ggallus.ENSEMBL.galGal6))

TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 106)
genes <- trim(genes(TxDb_galGal6))
exons <- trim(exons(TxDb_galGal6))
tss <- trim(promoters(TxDb_galGal6, upstream = 2000, downstream = 500))
gene_annotation <- SimpleList(genes, exons, tss)

# BUG This ends up being NULL because this fails
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, 
  genes = genes, exons = exons, TSS = tss,  annoStyle = "ENSEMBL")
arch_gg6_genes
arch_gg6_anno <- createGenomeAnnotation(genome = BSgenome.Ggallus.ENSEMBL.galGal6, filterChr = "MT")
```

Get Input Fragment Files

The first thing we need to do is separate out the scATAC-seq input files from the scRNA-seq input files.
The important thing is to end up with a named vector of files corresponding to the scATAC-seq data and a named vector of files corresponding to the scRNA-seq data
 __and that the names of the corresponding samples match__! 

```{r}
prefix <- "/workspaces/NC_scATAC/NC_Multiome/outs/"
input_files <- c(paste0(prefix,"atac_possorted_bam.bam"), paste0(prefix,"nc_filtered_feature_bc_matrix.h5"))
names(input_files) <- c("NC_Multiome","NC_Multiome")


# FEAT preflight check that if using atac_possorted from input_file that bcTag is not NULL
# BUG blocking geneAnnotation is some error with selecting keys from the GeneAnnotation in createGenomeAnnotation
arrow_files <- createArrowFiles(
  inputFiles = input_files[1],
  outputNames = c("NC_Multiome"),
  geneAnnotation = arch_gg6_genes,
  genomeAnnotation = arch_gg6_anno,
  bcTag = c("CB"),
  gsubExpression = c("atac_possorted#|gex#"),
  minTSS = 0.15,
  minFrags = 3500,
  maxFrags = 500000,  # maybe 500K even
  promoterRegion = c(2000, 500),
  addGeneScoreMat = FALSE,
  excludeChr = c("MT"),
  addTileMat = TRUE,
  threads = 48,
  cleanTmp = FALSE,
  force = TRUE
)

#ArchRProject
proj <- ArchRProject( 
  ArrowFiles = "NC_Multiome.arrow",
  outputDirectory = "ArchROutput",
  copyArrows = TRUE,
  geneAnnotation = arch_gg6_genes,
  genomeAnnotation = arch_gg6_anno,
  showLogo = TRUE,
  threads = getArchRThreads())

#   SAVE POINT HERE, USED FOR PRE-FILTERING PLOTS.
# saveArchRProject(proj, outputDirectory = "ArchROutput")
# proj <- readArchRProject("ArchROutput")
```

```{r}
#Import scRNA
seRNA <- import10xFeatureMatrix(
  input = input_files[2],
  names = c("NC_Multiome"),
  strictMatch = TRUE
  )

proj <- addGeneExpressionMatrix(input = proj, seRNA = seRNA, excludeChr = "MT", chromSizes = arch_gg6_anno$chromSizes)

# Doublet Filtration. Currently disabled just for tutorial. If you want to filter doublets uncomment below.
proj <- addDoubletScores(proj)
proj <- filterDoublets(proj)
```


## Naive analysis

Now that we have performed basic QC filtering and doublet removal

Filtering 2272 cells from ArchRProject!
        NC_Multiome : 2272 of 18739 (12.1%)

We can perform a niave analysis of the data to see what we can find. Afterwards, we have a couple of different options
for integration with existing analysis.

1. Add a new peakset (from existing timecourse) to the ArchR project and re-quantify to align the ATAC-Seq data.
  Next, map clusters from the existing timecourse to the new data. Then, using the new GEX data, call differential expression based on these clusters.
2. Perform some sort of integration (Harmony) across the ATAC samples and repeat all analysis with additional ATAC samples. 
  Then, use the shared cell barcodes to map gene expression to the ATAC tree.
3. Identify clusters in the new data and overlay cluster peak enrichment from the timecourse to identity lineages and confirm with independent GEX.

```{r}
# DEBUG
# Inspect SE matrix with
se <- getGroupSE(proj, useMatrix = "GeneExpressionMatrix")
se <- getGroupSE(proj, useMatrix = "TileMatrix")
#LSI-ATAC
proj <- addIterativeLSI(
  ArchRProj = proj,
  clusterParams = list(
    resolution = 0.2,
    sampleCells = 10000,
    n.start = 10
  ),
  saveIterations = FALSE,
  useMatrix = "TileMatrix",
  depthCol = "nFrags",
  name = "LSI_ATAC"
)

#LSI-RNA
proj <- addIterativeLSI(
  ArchRProj = proj,
  clusterParams = list(
    resolution = 0.2,
    sampleCells = 10000,
    n.start = 10
  ),
  saveIterations = FALSE,
  useMatrix = "GeneExpressionMatrix",
  depthCol = "Gex_nUMI",
  varFeatures = 2500,
  firstSelection = "variable",
  binarize = FALSE,
  name = "LSI_RNA"
)

#Combined Dims
proj <- addCombinedDims(proj, reducedDims = c("LSI_ATAC", "LSI_RNA"), name =  "LSI_Combined")

#UMAPs
proj <- addUMAP(proj, reducedDims = "LSI_ATAC", name = "UMAP_ATAC", minDist = 0.8, force = TRUE)
proj <- addUMAP(proj, reducedDims = "LSI_RNA", name = "UMAP_RNA", minDist = 0.8, force = TRUE)
proj <- addUMAP(proj, reducedDims = "LSI_Combined", name = "UMAP_Combined", minDist = 0.8, force = TRUE)
#Add Clusters
proj <- addClusters(proj, reducedDims = "LSI_Combined", name = "Clusters", resolution = 0.4, force = TRUE)

#Plot Embedding
p1 <- plotEmbedding(projMulti2, name = "Clusters_Combined", embedding = "UMAP_ATAC", size = 1, labelAsFactors=F, labelMeans=F)
p2 <- plotEmbedding(projMulti2, name = "Clusters_Combined", embedding = "UMAP_RNA", size = 1, labelAsFactors=F, labelMeans=F)
p3 <- plotEmbedding(projMulti2, name = "Clusters_Combined", embedding = "UMAP_Combined", size = 1, labelAsFactors=F, labelMeans=F)

p <- lapply(list(p1,p2,p3), function(x){
  x + guides(color = "none", fill = "none") + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0.1, 0.1, 0.1, 0.1), "cm")) +
    theme(
      axis.text.x=element_blank(), 
      axis.ticks.x=element_blank(), 
      axis.text.y=element_blank(), 
      axis.ticks.y=element_blank()
    )
})

do.call(cowplot::plot_grid, c(list(ncol = 3),p))
plotPDF(p1, p2, p3, name = "NC_Multiome_UMAP-scATAC-scRNA-Combined", addDOC = FALSE)
```

## Load Existing ArchR scATAC seq data

We also want to integrate these new samples (HH6 and HH18) with the existing timecourse based upon their scATAC seq signal.
Optimally, we quantify the signal in the exact peakset used in the original analysis. In this first round, we will used



```{r}
# TODO we should copy this to a local directory
# projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_final_thesis/")
# projNC@projectMetadata$outputDirectory <- "./Figure_Exports"
# projNC <- addImputeWeights(projNC, sampleCells = 10000, nRep = 3, k = 20)
# cluster_to_identity <- data.frame(cluster = paste0("C",1:10),
#                                   identity = c("C1_Melanocytes","C2_Sensory_Glia","C3_Int_Mesenchyme",
#                                                "C4_Multipotent_NC","C5_Hindbrain_Neurons","C6_Somatosensory_Neurons",
#                                                "C7_Midbrain_Neurons","C8_Early_Mesenchyme","C9_Sensory_Neurons",
#                                                "C10_Late_Mesenchyme"))
# projNC$Cell_Identity <- lookup(projNC$Clusters, cluster_to_identity)
```