---
title: "NC Multiome"
output: html_notebook
---

# NC Multiome

## Setup


```{r}
library(renv)
renv::activate()
# install.packages("BSgenome.Ggallus.ENSEMBL.galGal6_1.0.3.tar.gz", repos = NULL)

suppressPackageStartupMessages({
  library(ArchR)
  library(BSgenome.Ggallus.ENSEMBL.galGal6)
  library(GenomicFeatures)
  library(RMariaDB)
  library(org.Gg.eg.db)
})

addArchRThreads(threads = 44)
set.seed(1)
```

Genome Annotation

```{r}
# Make a blacklist for AADN05001427.1
# Manually remove noncontigous chromosomes.
# TxDb_galGal6 <- makeTxDbFromGFF("/workspaces/NC_scATAC/genes.gtf.gz",
#   organism = "Gallus gallus", dataSource = "Ensembl", chrominfo=seqinfo(BSgenome.Ggallus.ENSEMBL.galGal6))

TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 106)
genes <- trim(genes(TxDb_galGal6))
exons <- trim(exons(TxDb_galGal6))
tss <- trim(promoters(TxDb_galGal6, upstream = 2000, downstream = 500))
gene_annotation <- SimpleList(genes, exons, tss)
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, 
  genes = genes, exons = exons, TSS = tss,  annoStyle = "ENSEMBL")
arch_gg6_genes
arch_gg6_anno <- createGenomeAnnotation(genome = BSgenome.Ggallus.ENSEMBL.galGal6, filterChr = "MT")
```

Get Input Fragment Files

The first thing we need to do is separate out the scATAC-seq input files from the scRNA-seq input files.
The important thing is to end up with a named vector of files corresponding to the scATAC-seq data and a named vector of files corresponding to the scRNA-seq data
 __and that the names of the corresponding samples match__! 

```{r, eval=FALSE}
prefix <- "/workspaces/NC_scATAC/NC_Multiome/outs/"
input_files <- c(paste0(prefix,"atac_possorted_bam.bam"), paste0(prefix,"nc_filtered_feature_bc_matrix.h5"))
names(input_files) <- c("NC_Multiome", "NC_Multiome")

# addGebeExpressionMatrix has a bug in it, but we overwrite it with the correct one anyways
arrow_files <- createArrowFiles(
  inputFiles = input_files[1],
  outputNames = c("NC_Multiome"),
  geneAnnotation = arch_gg6_genes,
  genomeAnnotation = arch_gg6_anno,
  bcTag = c("CB"),
  gsubExpression = c("NC_Multiome#"),
  minTSS = 0.15,
  minFrags = 3500,
  maxFrags = 500000,  # maybe 500K even
  promoterRegion = c(2000, 500),
  addGeneScoreMat = FALSE,
  excludeChr = c("MT"),
  addTileMat = TRUE,
  cleanTmp = FALSE,
  force = TRUE
)
```

```{r}
#ArchRProject
proj <- ArchRProject(
  ArrowFiles = "NC_Multiome.arrow",
  outputDirectory = "ArchROutput",
  copyArrows = TRUE,
  geneAnnotation = arch_gg6_genes,
  genomeAnnotation = arch_gg6_anno,
  showLogo = TRUE,
  threads = getArchRThreads()
  )

# Import scRNA
seRNA <- import10xFeatureMatrix(
  input = input_files[2],
  names = c("NC_Multiome"),
  strictMatch = TRUE)

length(which(getCellNames(proj) %ni% colnames(seRNA)))

cells_to_remove <- which(getCellNames(proj) %ni% colnames(seRNA))
cells_to_keep <- which(getCellNames(proj) %in% colnames(seRNA))
cells <- getCellNames(proj)[cells_to_keep]
cells <- cells[-cells_to_remove]

proj <- subsetArchRProject(ArchRProj = proj,  cells = cells, outputDirectory = "NC_Multome_filt")
proj <- addGeneExpressionMatrix(input = proj, seRNA = seRNA, excludeChr = "MT", chromSizes = arch_gg6_anno$chromSizes)

# Doublet Filtration.
proj <- addDoubletScores(proj)
proj <- filterDoublets(proj)
```


## Naive analysis

Now that we have performed basic QC filtering and doublet removal

Filtering 2272 cells from ArchRProject!
        NC_Multiome : 2272 of 18739 (12.1%)

We can perform a niave analysis of the data to see what we can find. Afterwards, we have a couple of different options
for integration with existing analysis.

1. Add a new peakset (from existing timecourse) to the ArchR project and re-quantify to align the ATAC-Seq data.
  Next, map clusters from the existing timecourse to the new data. Then, using the new GEX data, call differential expression based on these clusters.
2. Perform some sort of integration (Harmony) across the ATAC samples and repeat all analysis with additional ATAC samples. 
  Then, use the shared cell barcodes to map gene expression to the ATAC tree.
3. Identify clusters in the new data and overlay cluster peak enrichment from the timecourse to identity lineages and confirm with independent GEX.

```{r}
# DEBUG
# Inspect SE matrix with
# se <- getGroupSE(proj, useMatrix = "TileMatrix")
proj <- loadArchRProject("NC_Multome_save")

#LSI-ATAC
proj <- addIterativeLSI(
  ArchRProj = proj,
  clusterParams = list(
    resolution = 0.2,
    sampleCells = 10000,
    n.start = 10
  ),
  saveIterations = FALSE,
  useMatrix = "TileMatrix",
  depthCol = "nFrags",
  name = "LSI_ATAC"
)

#LSI-RNA
proj <- addIterativeLSI(
  ArchRProj = proj,
  clusterParams = list(
    resolution = 0.2,
    sampleCells = 10000,
    n.start = 10
  ),
  saveIterations = FALSE,
  useMatrix = "GeneExpressionMatrix",
  depthCol = "Gex_nUMI",
  varFeatures = 2500,
  firstSelection = "variable",
  binarize = FALSE,
  name = "LSI_RNA"
)


saveArchRProject(proj, outputDirectory = "NC_Multome_save")
proj <- loadArchRProject("NC_Multome_save")

#Combined Dims
"""
Error in addCombinedDims2(proj, name = "LSI_Combined", reducedDims = c("LSI_ATAC",  : 
Matrices have different numbers of rows: 15847, 14, 15845, 15
"""
source("MultiModal.R")

proj <- addCombinedDims2(proj, name =  "LSI_Combined",
  reducedDims = c("LSI_ATAC", "LSI_RNA"),
  dimWeights = c(1,1),
  dimsToUse = c(1:20),
  scaleDims = F)
# 
# proj <- addCombinedDims(proj, name =  "LSI_Combined", reducedDims = c("LSI_ATAC", "LSI_RNA"), dimsToUse = c(1:30))

#UMAPs
proj <- addUMAP(proj, reducedDims = "LSI_ATAC", name = "UMAP_ATAC", minDist = 0.8, force = TRUE)
proj <- addUMAP(proj, reducedDims = "LSI_RNA", name = "UMAP_RNA", minDist = 0.8, force = TRUE)
proj <- addUMAP(proj, reducedDims = "LSI_Combined", name = "UMAP_Combined", minDist = 0.8, force = TRUE)
#Add Clusters
proj <- addClusters(proj, reducedDims = "LSI_Combined", name = "Clusters", resolution = 0.4, force = TRUE)

#Plot Embedding
p1 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP_ATAC", size = 1, labelAsFactors=F, labelMeans=F)
p2 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP_RNA", size = 1, labelAsFactors=F, labelMeans=F)
p3 <- plotEmbedding(proj, name = "Clusters", embedding = "UMAP_Combined", size = 1, labelAsFactors=F, labelMeans=F)

p <- lapply(list(p1,p2,p3), function(x){
  x + guides(color = "none", fill = "none") + 
    theme_ArchR(baseSize = 6.5) +
    theme(plot.margin = unit(c(0.1, 0.1, 0.1, 0.1), "cm")) +
    theme(
      axis.text.x=element_blank(),
      axis.ticks.x=element_blank(),
      axis.text.y=element_blank(),
      axis.ticks.y=element_blank()
    )
})

do.call(cowplot::plot_grid, c(list(ncol = 3),p))
plotPDF(p1, p2, p3, name = "NC_Multiome_UMAP-scATAC-scRNA-Combined", addDOC = FALSE)

### Subset into HH6_2 and HH18_2 based on initial clustering and my best guess.
cells_in_HH6_2 <- getCellNames(proj)[proj$Clusters %in% c("C9", "C8", "C1")]
cells_in_HH18_2 <- getCellNames(proj)[proj$Clusters %in% c("C2", "C3", "C4", "C5", "C6", "C7")]

sum(length(cells_in_HH6_2) + length(cells_in_HH18_2))

hh6_2 <- subsetArchRProject(
  ArchRProj = proj,
  cells = cells_in_HH6_2,
  outputDirectory = "NC_Multiome_HH6_2",
  dropCells = TRUE,
  logFile = NULL,
  threads = getArchRThreads(),
  force = FALSE
)

hh18_2 <- subsetArchRProject(
  ArchRProj = proj,
  cells = cells_in_HH18_2,
  outputDirectory = "NC_Multiome_HH18_2",
  dropCells = TRUE,
  logFile = NULL,
  threads = getArchRThreads(),
  force = FALSE
)

# SAVE POINT HERE
# Start of 7_NC_Multiome_Integration.Rmd
saveArchRProject(proj, outputDirectory = "NC_Multome_save")
```

### Additional notes

Below doesn't work.
You'll notice that there are some differences between the cluster residence of cells in the scATAC-seq space and cells in the scRNA-seq space. We can visualize these differences using a confusion matrix.

```{r, collapse=TRUE}
# cM_atac_rna <- confusionMatrix(paste0(proj$Clusters_ATAC), paste0(proj$Clusters_RNA))
# cM_atac_rna <- cM_atac_rna / Matrix::rowSums(cM_atac_rna)

# library(pheatmap)
# p_atac_rna <- pheatmap::pheatmap(
#   mat = as.matrix(cM_atac_rna), 
#   color = paletteContinuous("whiteBlue"), 
#   border_color = "black"
# )
```

Nearly all of the downstream operations that you will want to do downstream are equivalent to what is shown throughout the manual for the scATAC-seq-only analyses so we wont go into them here. As an example, to get peak-to-gene links from multiome data, we could use the following code.

```bash
python3 -m venv .venv
source .venv/bin/activate
pip install macs3

# Run report to estimate -l -u and -c for downstream
#macs3 hmmratac --cutoff-analysis-only -b your.bam
# I'm not going to re-run peak calling with hmmratac, but I WANT TO.
#macs3 hmmratac -b test/yeast_500k_SRR1822137.bam -n hmmratac_yeast500k
```

```{r, collapse=TRUE}
proj <- addGroupCoverages(ArchRProj = proj, groupBy = "Clusters")

# Add the GRANGE peakset from the original timecourse

# proj <- addReproduciblePeakSet(ArchRProj = proj, groupBy = "Clusters",
#   pathToMacs2 = "/workspaces/NC_scATAC/.venv/bin/macs3")
proj <- addPeakMatrix(ArchRProj = proj)
proj <- addPeak2GeneLinks(ArchRProj = proj, reducedDims = "LSI_Combined", 
 useMatrix = "GeneExpressionMatrix")
p2g <- getPeak2GeneLinks(ArchRProj = proj)

p2g[[1]]
```

There are, of course, some aspects of the analysis which you should tweak when using multiome data. One such example is the `bias` argument to `getMarkerFeatures()` which can be tweaked to account for both scATAC-seq data quality (`"TSSEnrichment`) and read depth for both assays (`"log10(nFrags)"` for scATAC-seq and `"log10(Gex_nUMI)"` for scRNA-seq).

```{r, collapse=TRUE}
se <- getMarkerFeatures(ArchRProj = proj,
                        groupBy = "Clusters_Combined",
                        bias = c("TSSEnrichment", "log10(nFrags)", "log10(Gex_nUMI)"))

heatmap_gex <- plotMarkerHeatmap(
  seMarker = se, 
  cutOff = "FDR <= 0.01 & Log2FC >= 2",
  nLabel = 4,
  transpose = TRUE
)

draw(heatmap_gex, heatmap_legend_side = "bot", annotation_legend_side = "bot")
```


## Load Existing ArchR scATAC seq data

We also want to integrate these new samples (HH6 and HH18) with the existing timecourse based upon their scATAC seq signal.
Optimally, we quantify the signal in the exact peakset used in the original analysis. In this first round, we will used



```{r}
# TODO we should copy this to a local directory
# projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_final_thesis/")
# projNC@projectMetadata$outputDirectory <- "./Figure_Exports"
# projNC <- addImputeWeights(projNC, sampleCells = 10000, nRep = 3, k = 20)
# cluster_to_identity <- data.frame(cluster = paste0("C",1:10),
#                                   identity = c("C1_Melanocytes","C2_Sensory_Glia","C3_Int_Mesenchyme",
#                                                "C4_Multipotent_NC","C5_Hindbrain_Neurons","C6_Somatosensory_Neurons",
#                                                "C7_Midbrain_Neurons","C8_Early_Mesenchyme","C9_Sensory_Neurons",
#                                                "C10_Late_Mesenchyme"))
# projNC$Cell_Identity <- lookup(projNC$Clusters, cluster_to_identity)
```