---
title: "R Notebook"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
    fig_width: 8
    fig_height: 6
editor_options:
  chunk_output_type: inline
---


```{r Loading libraries and data, message=FALSE, warning=FALSE}
suppressPackageStartupMessages({
  library(ArchR)
  library(knitr)
  library(TFBSTools)
  library(ComplexHeatmap)
  library(ggplot2)
  library(ggrepel)
  library(qdapTools)
  library(parallel)
})


addArchRThreads(threads = 62) 
set.seed(1)
projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_final/")
projNC@projectMetadata$outputDirectory <- "./Figure_Exports"
projNC <- addImputeWeights(projNC, sampleCells = 10000, nRep = 3, k = 20)
cluster_to_identity <- data.frame(cluster = paste0("C",1:10),
                                  identity = c("C1_Melanocytes","C2_Sensory_Glia","C3_Int_Mesenchyme",
                                               "C4_Multipotent_NC","C5_Neural_2","C6_Int_Neural",
                                               "C7_Neural_1","C8_Early_Mesenchyme","C9_Sensory_Neurons",
                                               "C10_Late_Mesenchyme"))
projNC$Cell_Identity <- lookup(projNC$Clusters, cluster_to_identity)
```

# Figure 1 - Dimensionality Reduction & Clustering

## A
A picture of a HH16/18 chicken embryo electroporated with the TFAP2aE1 enhancer alongside a schematic of single-cell ATAC-Seq.

## B
A two-dimensional UMAP projection colored by original sample identity (timepoint).
```{r}
p1 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                    name = "Sample", plotAs = "points",
                    pal = ArchRPalettes$bear, size = 0.2, rastr = F)
plotPDF(p1, name = "Plot-UMAP-Sample.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p1
```

## C
A two-dimensional UMAP projection followed by clustering in ArchR.
```{r}
p2 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData", 
                    name = "Clusters", plotAs = "points", size = 0.2, rastr = F)
plotPDF(p2, name = "Plot-UMAP-Clusters.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p2

p3 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData", 
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(p3, name = "Plot-UMAP-Cell_Identity.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p3
```


## D
A customized plot of enriched chromVAR motifs in the UMAP projection.
```{r}
motifs <- c("Pou5f1..Sox2","RUNX1","ASCL1","MEIS2","MITF","TWIST1","GRHL2","OTX2")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_motifs <- getFeatures(projNC, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z <- paste0("z:",plotting_motifs)

# Problem with threads > 1...
motif_plot_nocutoff <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)

for (plot in seq(length(motif_plot_nocutoff))){
  plotPDF(motif_plot_nocutoff[[plot]], name = paste0("Plot-Motif-Zscores-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}

combined <- cowplot::plot_grid(motif_plot_nocutoff[[1]],motif_plot_nocutoff[[2]],
                               motif_plot_nocutoff[[4]],motif_plot_nocutoff[[3]],
                               motif_plot_nocutoff[[8]],motif_plot_nocutoff[[7]],
                               motif_plot_nocutoff[[10]],motif_plot_nocutoff[[9]], nrow = 2)

 
plotPDF(combined, name = "Plot-Motif-Zscores.pdf", ArchRProj = projNC, addDOC = FALSE, width = 16, height = 16)

combined
```
## E
A UMAP projection of the 'end' of each lineage
```{r}
end1 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C2_Sensory_Glia"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end1, name = "Plot-UMAP-SensoryGlia.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end1

end2 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C9_Sensory_Neurons"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end2, name = "Plot-UMAP-SensoryNeurons.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end2

end3 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C7_Neural_1"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end3, name = "Plot-UMAP-Neural_1.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end3

end4 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C5_Neural_2"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end4, name = "Plot-UMAP-Neural_2.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end4

end5 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C10_Late_Mesenchyme"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end5, name = "Plot-UMAP-Late_Mesenchyme.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end5
```


# Figure S1 - QC Metrics
We saved a pre-filtering version of our object, and use it here to show pre-filtering QC.

## Prep
```{r}
library(ArchR)
library(knitr)
library(TFBSTools)
library(ComplexHeatmap)
library(ggplot2)
library(BSgenome.Ggallus.ENSEMBL.galGal6)
library(org.Gg.eg.db)
library(GenomicFeatures)
library(dplyr)
library(qdapTools)

addArchRThreads(threads = 32) 
set.seed(1)

HH6_dir <- "/data/Austin/10XscATAC/HH6_10xscATAC_Lane1/outs/"
HH8_dir <- "/data/Austin/10XscATAC/HH8_10xscATAC_Lane1_Lane2/outs/"
HH10_dir <- "/data/Austin/10XscATAC/HH10_10xscATAC_Lane1_Lane2/outs/"
HH12_dir <- "/data/Austin/10XscATAC/HH12_10xscATAC_Lane1_Lane2/outs/"
HH14_dir <- "/data/Austin/10XscATAC/HH14_10xscATAC_Lane1_Lane2/outs/"
HH16_dir <- "/data/Austin/10XscATAC/HH16_10xscATAC_Lane1/outs/"
HH18_dir <- "/data/Austin/10XscATAC/HH18_10xscATAC_Lane1/outs/"
libraries <- c("HH6","HH8","HH10","HH12","HH14","HH16","HH18")

# Genome annotation
arch_gg6 <- createGenomeAnnotation(genome = BSgenome.Ggallus.ENSEMBL.galGal6, filter = F)
# Manually remove noncontigous chromosomes.
arch_gg6$chromSizes <- arch_gg6$chromSizes[1:35]
TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 103)
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, annoStyle = "ENSEMBL")

doublet_cells <- as.character(read.csv(file = "doublet_cells.csv", row.names = 1)[,1])

projNC_pre <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_Prefiltered/")

projNC_pre@projectMetadata$outputDirectory <- "./Figure_Exports"
projNC_pre$DoubletCall <- ifelse(projNC_pre$cellNames %in% doublet_cells,
                                 "Doublet","Singlet")
```

## A
A Fragment-size distribution plot 
```{r}
p4 <- plotFragmentSizes(ArchRProj = projNC_pre)
plotPDF(p4, name = "Prefiltered-Plot-Sample-Fragment-Dist.pdf", ArchRProj = projNC_pre,
        addDOC = FALSE, width = 8, height = 8)
p4
```

## B
A TSS enrichment plot and FRiP score plot
```{r}
p5 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 3.5)

plotPDF(p5, name = "Prefiltered-Plot-Sample-TSS.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p5

p6 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "FRIP",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 0.4)

plotPDF(p6, name = "Prefiltered-Plot-Sample-FRiP.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p6
```
## C 
Doublet Detection Plot
```{r}
p7 <- plotEmbedding(ArchRProj = projNC_pre, colorBy = "cellColData", name = "DoubletCall", embedding = "UMAP", size = 1, baseSize = 14, plotAs = "points")

plotPDF(p7, name = "Prefiltered-Plot-UMAP-Doublet.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)

p7
```

## D
Depth Plot
```{r}
p8 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 3.3)
plotPDF(p8, name = "Prefiltered-Plot-Sample-Depth.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p8
```

## E
Extra motif plots of markers
```{r}
motifs <- c("YY2","SIX3","TFAP2A","PAX6","FOXD3","NR2F2","RFX3")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_motifs <- getFeatures(projNC, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z <- paste0("z:",plotting_motifs)

# Problem with threads > 1...
motif_plot_supp <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)

for (plot in seq(length(motif_plot_supp))){
  plotPDF(motif_plot_supp[[plot]], name = paste0("Plot-Motif-Zscores-Supp-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}
combined2 <- cowplot::plot_grid(motif_plot_supp[[9]],motif_plot_supp[[3]],
                                motif_plot_supp[[8]],motif_plot_supp[[2]],
                                motif_plot_supp[[1]],motif_plot_supp[[6]],
                                motif_plot_supp[[7]], nrow = 2)

plotPDF(combined2, name = "Plot-Motif-Zscores-supp.pdf", ArchRProj = projNC, addDOC = FALSE, width = 16, height = 16)

combined2
```
## F
A heatmap of the top changing motifs
```{r}
#scales::show_col(unlist(ArchRPalettes))
Cluster_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

Motif_list <- getMarkers(Cluster_Motifs, cutOff = "FDR <= 1e-50")
Motif_df <- as.data.frame(Motif_list)
top_Motifs <- Motif_df %>% dplyr::group_by(group_name) %>% dplyr::top_n(30, wt = -log10(FDR))

# Motifs by cluster mean Z-Score?
temp <- Cluster_Motifs@assays@data$Mean
# Add rownames
rownames(temp) <- Cluster_Motifs@elementMetadata$name

temp_f <- temp[unique(top_Motifs$name),]

# Who do we want to label?

label_motifs <- c("TWIST1_741", "OTX2_718","SNAI2_729",
                  "Dmbx1_366","ZEB1_412","ZBTB18_192","Pou5f1..Sox2_38",
                  "RUNX1_41","ASCL1_594","MEIS2_262","MITF_707","GRHL2_693",
                  "YY2_744","SIX1_405","TFAP2A_739","PAX6_14","NR2F2_399",
                  "RFX3_725","SP4_177","FOXD2_604","Atoh1_306","NEUROG2_162",
                  "SCRT1_727","Rarg_344","PITX2_526","SOX9_20","PHOX2B_720","ETV4_682",
                  "KLF2_496","ELK1_357")

label_index <- which(rownames(temp_f) %in% label_motifs)
label_text <- toupper(stringr::str_split(rownames(temp_f)[label_index], pattern = "_", simplify = T)[,1])

heatmap1 <- ComplexHeatmap::Heatmap(matrix = temp_f, cluster_rows = T, 
                        cluster_columns = T,
                        show_row_names = F,
                        use_raster = F, col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

plotPDF(heatmap1, name = "Plot-Motif-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 10)

heatmap1
```

# Figure 2 - Pseudotime
## A
A two-dimensional UMAP projection colored by real timepoint-normalized pseudotime projections
```{r}
p9 <- plotTrajectory(projNC, trajectory = "hour_adj_pseudotime",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9[[1]], name = "Plot-UMAP-Pseudotime.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9[[1]]
```

## B
Motifs that change over pseudotime.
Highlighting the mesencyhme, neural 1, neural 2, sensory ganglia, and sensory neuron lineages.
First we have to isolate the cells of that lineage to perform the analysis on (differentially enriched).
Then, we aggregate transcription factor signal by family to reduce noise.
Finally, we plot a heatmap and label specific transcription factors of interest in each lineage.

```{r Family Aggregation}
matrix_agg_scores <- function(input = sm_mat, transpose = T) {
  ## A function to aggregate motif scores by family.
  
  if (transpose){
    input <- t(input)
  }
  # Convert to DF
  input <- as.data.frame(input)
  # Clean up names
  colnames(input) <- toupper(colnames(input))
  
  # GRHL
  if (any(c("GRHL1","GRHL2","GRHL3") %in% colnames(input))){
    print("Collapsing GRHL's")
    # Which are present?
    select <- c("GRHL1","GRHL2","GRHL3")[c("GRHL1","GRHL2","GRHL3") %in% colnames(input)]
    GRHL_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$GRHL_Family <- GRHL_Family
  }
  # TEAD
  if (any(c("TEAD1","TEAD2","TEAD3","TEAD4") %in% colnames(input))){
    print("Collapsing TEAD's")
    # Which are present?
    select <- c("TEAD1","TEAD2","TEAD3","TEAD4")[c("TEAD1","TEAD2","TEAD3","TEAD4") %in% colnames(input)]
    TEAD_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$TEAD_Family <- TEAD_Family
  }
  
  # TFAP2
  if (any(grep(pattern = "TFAP2", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing TAFP2's")
    select <- grep(pattern = "TFAP2", colnames(input), value = T)[grep(pattern = "TFAP2", colnames(input), value = T) %in% colnames(input)]
    TFAP2_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$TFAP2_Family <- TFAP2_Family
  }  
  
  # SANI
  if (any(c("SNAI1", "SNAI2","SNAI3") %in% colnames(input))){
    print("Collapsing SANI's")
    # Which are present?
    select <- c("SNAI1", "SNAI2","SNAI3")[c("SNAI1", "SNAI2","SNAI3") %in% colnames(input)]
    SANI_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SANI_Family <- SANI_Family
  }   
  
  # SOXB
  if (any(c("SOX1", "SOX2","SOX3") %in% colnames(input))){
    print("Collapsing SOXB's")
    # Which are present?
    select <- c("SOX1", "SOX2","SOX3")[c("SOX1", "SOX2","SOX3") %in% colnames(input)]
    SOXB_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXB_Family <- SOXB_Family
  }
  
  # SOXD
  if (any(c("SOX5", "SOX6","SOX13") %in% colnames(input))){
    print("Collapsing SOXD's")
    # Which are present?
    select <- c("SOX5", "SOX6","SOX13")[c("SOX5", "SOX6","SOX13") %in% colnames(input)]
    SOXD_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXD_Family <- SOXD_Family
  }
  
  # SOXE
  if (any(c("SOX8", "SOX9","SOX10") %in% colnames(input))){
    print("Collapsing SOXE's")
    # Which are present?
    select <- c("SOX8", "SOX9","SOX10")[c("SOX8", "SOX9","SOX10") %in% colnames(input)]
    SOXE_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXE_Family <- SOXE_Family
  }
  
  # SOXF
  if (any(c("SOX7", "SOX17","SOX18") %in% colnames(input))){
    print("Collapsing SOXF's")
    # Which are present?
    select <- c("SOX7", "SOX17","SOX18")[c("SOX7", "SOX17","SOX18") %in% colnames(input)]
    SOXF_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXF_Family <- SOXF_Family
  }
  
  # NR2F
  if (any(grep(pattern = "NR2F", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing NR2F's")
    select <- grep(pattern = "NR2F", colnames(input), value = T)[grep(pattern = "NR2F", colnames(input), value = T) %in% colnames(input)]
    NR2F_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$NR2F_Family <- NR2F_Family
  } 
  
  # SP
  if (any(c("SP1", "SP2","SP3","SP4","SP8","SP9") %in% colnames(input))){
    print("Collapsing SP's")
    # Which are present?
    select <- c("SP1", "SP2","SP3","SP4","SP8","SP9")[c("SP1", "SP2","SP3","SP4","SP8","SP9") %in% colnames(input)]
    SP_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SP_Family <- SP_Family
  }
  
  # FOX
  if (any(grep(pattern = "FOX", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing FOX's")
    select <- grep(pattern = "FOX", colnames(input), value = T)[grep(pattern = "FOX", colnames(input), value = T) %in% colnames(input)]
    FOX_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$FOX_Family <- FOX_Family
  } 
  
  # KLF
  if (any(grep(pattern = "KLF", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing KLF's")
    select <- grep(pattern = "KLF", colnames(input), value = T)[grep(pattern = "KLF", colnames(input), value = T) %in% colnames(input)]
    KLF_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$KLF_Family <- KLF_Family
  } 
  
  # RFX
  if (any(grep(pattern = "RFX", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing RFX's")
    select <- grep(pattern = "RFX", colnames(input), value = T)[grep(pattern = "RFX", colnames(input), value = T) %in% colnames(input)]
    RFX_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$RFX_Family <- RFX_Family
  } 
  
  # RXR
  if (any(grep(pattern = "RFR", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing RFR's")
    select <- grep(pattern = "RFR", colnames(input), value = T)[grep(pattern = "RFR", colnames(input), value = T) %in% colnames(input)]
    RFR_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$RFR_Family <- RFR_Family
  } 
  
  # SIX
  if (any(grep(pattern = "SIX", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing SIX's")
    select <- grep(pattern = "SIX", colnames(input), value = T)[grep(pattern = "SIX", colnames(input), value = T) %in% colnames(input)]
    SIX_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SIX_Family <- SIX_Family
  }
  
  # ETV
  if (any(grep(pattern = "ETV", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing ETV's")
    select <- grep(pattern = "ETV", colnames(input), value = T)[grep(pattern = "ETV", colnames(input), value = T) %in% colnames(input)]
    ETV_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$ETV_Family <- ETV_Family
  }
  
  return(input)
}

```

```{r Mesenchyme TF over pseudotime}
# Cell names along mesenchyme trajectory
mes_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C3","C8","C10")]

which(projNC$cellNames %in% mes_cells)
# Motif scores per cell
motif_mat <- getMatrixFromProject(projNC,
                                  useMatrix = "MotifMatrix",
                                  useSeqnames = "z")
motif_mat_raw <- assay(motif_mat)
motif_mat_impute <- imputeMatrix(mat = motif_mat_raw,
                                 imputeWeights = getImputeWeights(projNC))
# Order columns by pseudotime.
mes_cells_df <- data.frame(Name = mes_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in% c("C4","C3","C8","C10")])
mes_cells_df <- mes_cells_df[order(mes_cells_df$Pseudotime),]

motif_mat_impute_ord <- as.matrix(motif_mat_impute[,mes_cells_df$Name])
# Collapse by motif family.
rownames(motif_mat_impute_ord) <- stringr::str_split(rownames(motif_mat_impute_ord), pattern = "_", simplify = T)[,1]

motif_mat_impute_ord <- t(matrix_agg_scores(motif_mat_impute_ord))

# Filter to only enriched TF's
mes_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C3","C8","C10"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

mes_motif_df <- as.data.frame(getMarkers(mes_Motifs, cutOff = "FDR <= 1e-20 & MeanDiff > 2"))

selected_motifs <- toupper(unique(mes_motif_df$name))
selected_motifs <- stringr::str_split(selected_motifs, pattern = "_", simplify = T)[,1]

# Remove motifs NOT in the family-collapsed matrix.
selected_motifs <- selected_motifs[selected_motifs %in% rownames(motif_mat_impute_ord)]

# Add in the family names
selected_motifs <- c(selected_motifs, grep("_Family", rownames(motif_mat_impute_ord), value = T))

labels <- c("ETS1","RXRG","NR2F_Family","TWIST1",
            "ZBTB18","TFAP2_Family","JUNB",
            "FOSL1..JUN","FLI1","SOXE_Family")

labels[!labels %in% selected_motifs]

label_index <- which(selected_motifs %in% labels)


heatmap3 <- ComplexHeatmap::Heatmap(
  matrix = motif_mat_impute_ord[selected_motifs,],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Mesenchyme",
  col = as.character(ArchRPalettes$blueYellow)) +
  rowAnnotation(names = anno_mark(at = label_index, labels = labels))

selected_motifs[row_order(heatmap3)]

plotPDF(heatmap3, name = "Plot-Mesenchyme-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 5)

heatmap3
```

```{r Neural 1 TF over pseudotime}
# Cell names along Neural 1 trajectory
neural1_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C3","C5")]

# Order columns by pseudotime.
neural1_cells_df <- data.frame(Name = neural1_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C3","C5")])

neural1_cells_df <- neural1_cells_df[order(neural1_cells_df$Pseudotime),]

ordered_neural1 <- as.matrix(motif_mat_impute[,neural1_cells_df$Name])
# Collapse by motif family.
rownames(ordered_neural1) <- stringr::str_split(rownames(ordered_neural1), pattern = "_", simplify = T)[,1]

ordered_neural1 <- t(matrix_agg_scores(ordered_neural1))

# Filter to only enriched TF's
neural1_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C4","C3","C5"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

neural1_motif_df <- as.data.frame(getMarkers(neural1_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1"))

selected_motifs <- toupper(unique(neural1_motif_df$name))
selected_motifs <- stringr::str_split(selected_motifs, pattern = "_", simplify = T)[,1]

# Remove motifs NOT in the family-collapsed matrix.
selected_motifs <- selected_motifs[selected_motifs %in% rownames(ordered_neural1)]

# Add in the family names
selected_motifs <- c(selected_motifs, grep("_Family", rownames(ordered_neural1), value = T))

# Choose which transcription factors to highlight
labels <- c("POU5F1..SOX2","RFX_Family","MEIS1.VAR.2","SOXB_Family",
            "PBX1","MEIS2","ZIC_Family","PBX2"
            )

# Ensure labels are in selected motifs
labels[!labels %in% selected_motifs]
label_index <- which(selected_motifs %in% labels)
label_text <- toupper(stringr::str_split(rownames(ordered_neural1[selected_motifs,])[label_index], pattern = "_", simplify = T)[,1])

heatmap4 <- ComplexHeatmap::Heatmap(
  matrix = ordered_neural1[selected_motifs,],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Neural 1",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(neural1_motif_df$name)[row_order(heatmap4)]


plotPDF(heatmap4, name = "Plot-neural1-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 5)

heatmap4
```

```{r Neural 2 TF over pseudotime}
# Cell names along Neural 2 trajectory
neural2_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C6","C7")]

which(projNC$cellNames %in% neural2_cells)

# Order columns by pseudotime.
neural2_cells_df <- data.frame(Name = neural2_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C6","C7")])

neural2_cells_df <- neural2_cells_df[order(neural2_cells_df$Pseudotime),]

ordered_neural2 <- as.matrix(motif_mat_impute[,neural2_cells_df$Name])

# Filter to only enriched TF's
neural2_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C4","C6","C7"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

neural2_motif_df <- as.data.frame(getMarkers(neural2_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("OTX2_718","Pou5f1..Sox2_38","Dmbx1_366",
            "PITX2_526","SOX2_730","Arid3a_43","PHOX2A_204",
            "Shox2_211","GBX1_372","GATA1_690")

label_index <- which(unique(neural2_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(neural2_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap5 <- ComplexHeatmap::Heatmap(
  matrix = ordered_neural2[unique(neural2_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Neural 2",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(neural2_motif_df$name)[row_order(heatmap5)]


plotPDF(heatmap5, name = "Plot-neural2-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 5)

heatmap5
```

```{r Sensory glia TF over pseudotime}
# Cell names along sensory glia trajectory
sensory_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C2")]

which(projNC$cellNames %in% sensory_cells)

# Order columns by pseudotime.
sensory_cells_df <- data.frame(Name = sensory_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C2")])

sensory_cells_df <- sensory_cells_df[order(sensory_cells_df$Pseudotime),]

ordered_sensory <- as.matrix(motif_mat_impute[,sensory_cells_df$Name])

# Filter to only enriched TF's
sensory_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C2"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

sensory_motif_df <- as.data.frame(getMarkers(sensory_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("GRHL2_693","ASCL1.var.2_639","ZEB1_412",
            "SNAI2_729","SIX1_405","DLX5_464",
            "MEOX2_196","ISL2_388","TEAD4_738",
            "GATA2_390")

label_index <- which(unique(sensory_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(sensory_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap6 <- ComplexHeatmap::Heatmap(
  matrix = ordered_sensory[unique(sensory_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Sensory",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(sensory_motif_df$name)[row_order(heatmap6)]


plotPDF(heatmap6, name = "Plot-sensory-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 5)

heatmap6
```

```{r Sensory neurons TF over pseudotime}
# Cell names along sensory neuron trajectory
sensory_neuron_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C9")]

which(projNC$cellNames %in% sensory_neuron_cells)

# Order columns by pseudotime.
sensory_neuron_cells_df <- data.frame(Name = sensory_neuron_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C9")])

sensory_neuron_cells_df <- sensory_neuron_cells_df[order(sensory_neuron_cells_df$Pseudotime),]

ordered_sensory <- as.matrix(motif_mat_impute[,sensory_neuron_cells_df$Name])

# Filter to only enriched TF's
sensory_neuron_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C9"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

sensory_neuron_motif_df <- as.data.frame(getMarkers(sensory_neuron_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("Foxd3_8","FOXE1_472","FOXO4_334","HAND2_646",
            "NEUROG2.var.2_650","ATOH1.var.2_455","NEUROD1_395",
            "Ascl2_305","GATA5_605")

label_index <- which(unique(sensory_neuron_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(sensory_neuron_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap7 <- ComplexHeatmap::Heatmap(
  matrix = ordered_sensory[unique(sensory_neuron_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Sensory",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(sensory_neuron_motif_df$name)[row_order(heatmap7)]

plotPDF(heatmap7, name = "Plot-sensory-neuron-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 5)

heatmap7
```

## C
Individual line plots of just top Motifs for Mesenchyme branch
```{r}

interesting_mesenchyme_factors <- c("ETS1","RXRG","NR2F_Family","TWIST1",
            "ZBTB18","TFAP2_Family","JUNB",
            "FOSL1..JUN","FLI1","SOXE_Family")

# Motif Matrix for a gene within a particular trajectory
# TWIST1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","TWIST1_741"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-TWIST1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# NR2F1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","NR2F1_220"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-NR2F1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# ZBTB18
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","ZBTB18_192"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-ZBTB18-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# RXRG
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","RXRG_341"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-RXRG-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# FLI1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","FLI1_133"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-FLI1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# TFAP2B
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","TFAP2B_300"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-TFAP2B-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

```

## D
Individual line plots of top genes in Motif and Genescore.
```{r}
# Highlighting Specific Motifs across each branch
# Three per branch.

# Sensory Glia Markers
glia_tf <- c("ASCL1.var.2_639","GRHL2_693","DLX5_464")
glia_genescore <- c("GRHL3","ASCL1","DLX5")

plotting_tfs <- getFeatures(projNC, select = paste(glia_tf, collapse="|"), useMatrix = "MotifMatrix")
plotting_genes <- getFeatures(projNC, select = paste(glia_genescore, collapse="|"), useMatrix = "GeneScoreMatrix")

# Let's see if ArchR fixed it's plotting of this yet...
# Works, but doesn't like multiple features at a time.

# Motif Matrix for a gene within a particular trajectory
p11 <- plotTrajectory(projNC, trajectory = "C2END", colorBy = "MotifMatrix",
                      name = paste0("z:","GRHL2_693"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
# Genescore Matrix for a gene within a particular trajectory 
p12 <- plotTrajectory(projNC, trajectory = "C2END", colorBy = "GeneScoreMatrix",
                      name = "GRHL3", addArrow = F, plotAs = "points",
                      rastr = F, continuousSet = "horizonExtra")

comb_umap <- ggAlignPlots(p11[[1]], p12[[1]], type = "h", draw = F)
plotPDF(comb_umap, name = "Plot-GRHL3-C2END-Motif-Genescore-UMAP.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 8, height = 6)
comb_line <- ggAlignPlots(p11[[2]], p12[[2]], type = "h", draw = F)
plotPDF(comb_line,name = "Plot-GRHL3-C2END-Motif-Genescore-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 8, height = 6)

```

Let's look at geneScores for some of the top motifs and compare them.

```{r}
motif_to_hi <- c("SOX8", "TWIST1","RFX2","ZIC1","OTX2","Dmbx1","YY2","RFX3","GRHL3","RFX1","RFX4")

plotting_motifs2 <- getFeatures(projNC, select = paste(motif_to_hi, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z2 <- paste0("z:",plotting_motifs2)


motif_highlights <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z2,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     randomize = F,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)

plotting_genes2 <- getFeatures(projNC, select = paste(motif_to_hi, collapse="|"), useMatrix = "GeneScoreMatrix")

gene_highlights <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "GeneScoreMatrix",
                                     name = plotting_genes2,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     randomize = F,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)


combined <- cowplot::plot_grid(motif_highlights[[1]],gene_highlights[[9]],
                               motif_highlights[[2]],gene_highlights[[5]],
                               ncol = 2)

combined2 <- cowplot::plot_grid(motif_highlights[[3]],gene_highlights[[1]],
                               motif_highlights[[4]],gene_highlights[[8]],
                               ncol = 2)

combined3 <- cowplot::plot_grid(motif_highlights[[6]],gene_highlights[[2]],
                               motif_highlights[[7]],gene_highlights[[7]],
                               ncol = 2)

combined4 <- cowplot::plot_grid(motif_highlights[[9]],gene_highlights[[10]],
                                motif_highlights[[10]],gene_highlights[[3]],
                               ncol = 2)

combined5 <- cowplot::plot_grid(motif_highlights[[11]],gene_highlights[[6]],
                               ncol = 2)

plotPDF(combined, name = "Plot-Motif-GeneScore-UMAP-Projection-1.pdf", ArchRProj = projNC, addDOC = FALSE, width = 12, height = 10)
plotPDF(combined2, name = "Plot-Motif-GeneScore-UMAP-Projection-2.pdf", ArchRProj = projNC, addDOC = FALSE, width = 12, height = 10)
plotPDF(combined3, name = "Plot-Motif-GeneScore-UMAP-Projection-3.pdf", ArchRProj = projNC, addDOC = FALSE, width = 12, height = 10)
plotPDF(combined4, name = "Plot-Motif-GeneScore-UMAP-Projection-4.pdf", ArchRProj = projNC, addDOC = FALSE, width = 12, height = 10)
plotPDF(combined4, name = "Plot-Motif-GeneScore-UMAP-Projection-5.pdf", ArchRProj = projNC, addDOC = FALSE, width = 12, height = 10)

# How does the entire RFX family look?
combined6 <- cowplot::plot_grid(motif_highlights[[3]],gene_highlights[[1]],
                                motif_highlights[[2]],gene_highlights[[5]],
                                motif_highlights[[9]],gene_highlights[[10]],
                               ncol = 2)
plotPDF(combined6, name = "Plot-Motif-GeneScore-UMAP-Projection-RFX.pdf", ArchRProj = projNC, addDOC = FALSE, width = 12, height = 10)
```
## E 
Integrative pseudo-time analyses.
Let's find the best correlated Gene to Motif scores in an unbaised way.
```{r}
# For each lineage, find the best correlated TF+Genes

for(lineage in c("C1END","C2END","C5END","C7END","C9END","C10END")){
    END_MM  <- getTrajectory(ArchRProj = projNC, name = lineage, useMatrix = "MotifMatrix", log2Norm = FALSE)
  END_GSM <- getTrajectory(ArchRProj = projNC, name = lineage, useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
  END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.4)

  END_trajGSM2 <- END_GSM[END_MM_corGSM_MM[[1]]$name2, ]
  END_trajMM2 <- END_MM[END_MM_corGSM_MM[[1]]$name1, ]
  trajCombined <- END_trajGSM2
  assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
  combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
  rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
  ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
  ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
  combined_cor_heatmap <- ht1 + ht2
  plotPDF(combined_cor_heatmap, name = paste0(lineage,"_TF-Gene_Correlation.pdf"), ArchRProj = projNC,
           addDOC = FALSE)
}
```


# Figure 3
Genescores

## A
Investigation of Genescores.
```{r}
# Genescores on non-imputed data.
markersGS <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "Clusters",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    normBy = "ReadsInPeaks",
    testMethod = "wilcoxon")

# Heatmap of all differentially enriched genes.
plotMarkerHeatmap(seMarker = markersGS,
                  cutOff = "FDR <= 0.01 & Log2FC >= 0.75")

```



## B
DiffTF-like correlation of TF expression to accessibility of peaks that contain their motif.

We can use this to identify transcription factors of interest in each lineage.
Interested in FOXD3 'biomodal' activity.

```{r}
seGroupMotif <- getGroupSE(ArchRProj = projNC, useMatrix = "MotifMatrix", groupBy = "Clusters")
seZ <- seGroupMotif[rowData(seGroupMotif)$seqnames=="z",]

rowData(seZ)$maxDelta <- lapply(seq_len(ncol(seZ)), function(x){
  rowMaxs(assay(seZ) - assay(seZ)[,x])
}) %>% Reduce("cbind", .) %>% rowMaxs

corGSM_MM <- correlateMatrices(
    ArchRProj = projNC,
    useMatrix1 = "GeneScoreMatrix",
    useMatrix2 = "MotifMatrix",
    reducedDims = "IterativeLSI")

corGSM_MM$maxDelta <- rowData(seZ)[match(corGSM_MM$MotifMatrix_name, rowData(seZ)$name), "maxDelta"]

corGSM_MM <- corGSM_MM[order(abs(corGSM_MM$cor), decreasing = TRUE), ]
corGSM_MM <- corGSM_MM[which(!duplicated(gsub("\\-.*","",corGSM_MM[,"MotifMatrix_name"]))), ]
corGSM_MM$TFRegulator <- "NO"

# Positive
corGSM_MM$TFRegulator[which(corGSM_MM$cor > 0.3 & corGSM_MM$padj < 0.1 & corGSM_MM$maxDelta > quantile(corGSM_MM$maxDelta, 0.7))] <- "Positive"
# Negative
corGSM_MM$TFRegulator[which(corGSM_MM$cor < -0.3 & corGSM_MM$padj < 0.1 & corGSM_MM$maxDelta > quantile(corGSM_MM$maxDelta, 0.7))] <- "Negative"

sort(corGSM_MM[corGSM_MM$TFRegulator=="Positive",1])

corGSM_MM$label <- ifelse(corGSM_MM$TFRegulator %in% c("Positive","Negative"), corGSM_MM$GeneScoreMatrix_name, "")

p13 <- ggplot(data.frame(corGSM_MM), aes(cor, maxDelta, color = TFRegulator, label = label)) +
  geom_point() + 
  theme_ArchR() +
  geom_vline(xintercept = 0, lty = "dashed") + 
  scale_color_manual(values = c("NO"="darkgrey", "Positive"="firebrick3", "Negative"="blue")) +
  xlab("Correlation To Gene Score") +
  ylab("Max TF Motif Delta") +
  scale_y_continuous(
    expand = c(0,0), 
    limits = c(0, max(corGSM_MM$maxDelta)*1.05)
  ) + geom_label_repel(max.overlaps = 20, box.padding = 0.25)

plotPDF(p13, name = "Plot-GeneScore-TF-Correlation.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 7)

p13
```


```{r}

```
