---
title: "R Notebook"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
    fig_width: 8
    fig_height: 6
editor_options:
  chunk_output_type: inline
---


```{r Loading libraries and data, message=FALSE, warning=FALSE}
suppressPackageStartupMessages({
  library(ArchR)
  library(knitr)
  library(TFBSTools)
  library(ComplexHeatmap)
  library(ggplot2)
  library(ggrepel)
  library(qdapTools)
  library(parallel)
})


addArchRThreads(threads = 62) 
set.seed(1)
projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_final_thesis/")
projNC@projectMetadata$outputDirectory <- "./Figure_Exports"
projNC <- addImputeWeights(projNC, sampleCells = 10000, nRep = 3, k = 20)
cluster_to_identity <- data.frame(cluster = paste0("C",1:10),
                                  identity = c("C1_Melanocytes","C2_Sensory_Glia","C3_Int_Mesenchyme",
                                               "C4_Multipotent_NC","C5_Hindbrain_Neurons","C6_Somatosensory_Neurons",
                                               "C7_Midbrain_Neurons","C8_Early_Mesenchyme","C9_Sensory_Neurons",
                                               "C10_Late_Mesenchyme"))
projNC$Cell_Identity <- lookup(projNC$Clusters, cluster_to_identity)
```

# Figure 1 - Dimensionality Reduction & Clustering

## A
A picture of a HH16/18 chicken embryo electroporated with the TFAP2aE1 enhancer alongside a schematic of single-cell ATAC-Seq.

## B
A two-dimensional UMAP projection colored by original sample identity (timepoint).
```{r}
p1 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                    name = "Sample", plotAs = "points",
                    pal = ArchRPalettes$bear, size = 0.2, rastr = F)
plotPDF(p1, name = "Plot-UMAP-Sample.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p1
```

## C
A two-dimensional UMAP projection followed by clustering in ArchR.
```{r}
p2 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData", 
                    name = "Clusters", plotAs = "points", size = 0.2, rastr = F)
plotPDF(p2, name = "Plot-UMAP-Clusters.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p2

p3 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData", 
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(p3, name = "Plot-UMAP-Cell_Identity.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p3
```


## D
A customized plot of enriched chromVAR motifs in the UMAP projection.
```{r}
motifs <- c("Pou5f1..Sox2","RUNX1","ASCL1","MEIS2","MITF","TWIST1","GRHL2","OTX2")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_motifs <- getFeatures(projNC, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z <- paste0("z:",plotting_motifs)

# Problem with threads > 1...
motif_plot_nocutoff <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)

for (plot in seq(length(motif_plot_nocutoff))){
  plotPDF(motif_plot_nocutoff[[plot]], name = paste0("Plot-Motif-Zscores-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}

combined <- cowplot::plot_grid(motif_plot_nocutoff[[1]],motif_plot_nocutoff[[2]],
                               motif_plot_nocutoff[[4]],motif_plot_nocutoff[[3]],
                               motif_plot_nocutoff[[8]],motif_plot_nocutoff[[7]],
                               motif_plot_nocutoff[[10]],motif_plot_nocutoff[[9]], nrow = 2)

 
plotPDF(combined, name = "Plot-Motif-Zscores.pdf", ArchRProj = projNC, addDOC = FALSE, width = 16, height = 16)

combined
```
## E
A UMAP projection of the 'end' of each lineage
```{r}
end1 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C2_Sensory_Glia"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end1, name = "Plot-UMAP-SensoryGlia.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end1

end2 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C9_Sensory_Neurons"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end2, name = "Plot-UMAP-SensoryNeurons.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end2

end3 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C7_Midbrain_Neurons"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end3, name = "Plot-UMAP-Midbrain_Neurons.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end3

end4 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C5_Hindbrain_Neurons"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end4, name = "Plot-UMAP-Hindbrain_Neurons.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end4

end5 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C10_Late_Mesenchyme"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end5, name = "Plot-UMAP-Late_Mesenchyme.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end5

end6 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                      highlightCells = projNC$cellNames[projNC$Cell_Identity == "C6_Somatosensory_Neurons"],
                    name = "Cell_Identity", plotAs = "points", size = 0.2, rastr = F)
plotPDF(end6, name = "Plot-UMAP-Somatosensory_Neurons.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

end6
```


# Figure S1 - QC Metrics
We saved a pre-filtering version of our object, and use it here to show pre-filtering QC.

## Prep
```{r}
library(ArchR)
library(knitr)
library(TFBSTools)
library(ComplexHeatmap)
library(ggplot2)
library(BSgenome.Ggallus.ENSEMBL.galGal6)
library(org.Gg.eg.db)
library(GenomicFeatures)
library(dplyr)
library(qdapTools)

addArchRThreads(threads = 32) 
set.seed(1)

HH6_dir <- "/data/Austin/10XscATAC/HH6_10xscATAC_Lane1/outs/"
HH8_dir <- "/data/Austin/10XscATAC/HH8_10xscATAC_Lane1_Lane2/outs/"
HH10_dir <- "/data/Austin/10XscATAC/HH10_10xscATAC_Lane1_Lane2/outs/"
HH12_dir <- "/data/Austin/10XscATAC/HH12_10xscATAC_Lane1_Lane2/outs/"
HH14_dir <- "/data/Austin/10XscATAC/HH14_10xscATAC_Lane1_Lane2/outs/"
HH16_dir <- "/data/Austin/10XscATAC/HH16_10xscATAC_Lane1/outs/"
HH18_dir <- "/data/Austin/10XscATAC/HH18_10xscATAC_Lane1/outs/"
libraries <- c("HH6","HH8","HH10","HH12","HH14","HH16","HH18")

# Genome annotation
arch_gg6 <- createGenomeAnnotation(genome = BSgenome.Ggallus.ENSEMBL.galGal6, filter = F)
# Manually remove noncontigous chromosomes.
arch_gg6$chromSizes <- arch_gg6$chromSizes[1:35]
TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 103)
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, annoStyle = "ENSEMBL")

# TODO recover doublets.
doublet_cells <- as.character(read.csv(file = "doublet_cells.csv", row.names = 1)[,1])

projNC_pre <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_Prefiltered/")

projNC_pre@projectMetadata$outputDirectory <- "./Figure_Exports"
projNC_pre$DoubletCall <- ifelse(projNC_pre$cellNames %in% doublet_cells,
                                 "Doublet","Singlet")
```

## A
A Fragment-size distribution plot 
```{r}
p4 <- plotFragmentSizes(ArchRProj = projNC_pre)
plotPDF(p4, name = "Prefiltered-Plot-Sample-Fragment-Dist.pdf", ArchRProj = projNC_pre,
        addDOC = FALSE, width = 8, height = 8)
p4
```

## B
A TSS enrichment plot and FRiP score plot
```{r}
p5 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 3.5)

plotPDF(p5, name = "Prefiltered-Plot-Sample-TSS.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p5

p6 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "FRIP",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 0.4)

plotPDF(p6, name = "Prefiltered-Plot-Sample-FRiP.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p6
```
## C 
Doublet Detection Plot
```{r}
p7 <- plotEmbedding(ArchRProj = projNC_pre, colorBy = "cellColData", name = "DoubletCall", embedding = "UMAP", size = 1, baseSize = 14, plotAs = "points")

plotPDF(p7, name = "Prefiltered-Plot-UMAP-Doublet.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)

p7
```

## D
Depth Plot
```{r}
p8 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 3.3)
plotPDF(p8, name = "Prefiltered-Plot-Sample-Depth.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p8
```

## E
Extra motif plots of markers
```{r}
motifs <- c("YY2","SIX3","TFAP2A","PAX6","FOXD3","NR2F2","RFX3", "ZIC1")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_motifs <- getFeatures(projNC, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z <- paste0("z:",plotting_motifs)

# Problem with threads > 1...
motif_plot_supp <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1,
                                    randomize = F)

for (plot in seq(length(motif_plot_supp))){
  plotPDF(motif_plot_supp[[plot]], name = paste0("Plot-Motif-Zscores-Supp-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}

# Digging into neuronal clusters.

# Class B dILA Neurons
# LHX1/5 PAX2 DRGX LBX2 GBX1 EN1
neuron_motifs <- c("LBX2","TLX2","LHX5","GBX1")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_neuron_motifs <- getFeatures(projNC, select = paste(neuron_motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_neuron_motifs_z <- paste0("z:",plotting_neuron_motifs)

# Problem with threads > 1...
motif_plot_neurons <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_neuron_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1,
                                     randomize = F)
motif_plot_neurons

for (plot in seq(length(motif_plot_neurons))){
  plotPDF(motif_plot_neurons[[plot]], name = paste0("Plot-Motif-Zscores-C6-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}

# tbd neural 1
# dopaminergic neurons of substantia nigra?
# OTX2 RFX4 DMBX1 PITX3 RHOXF1

neuron_motifs_2 <- c("DMBX1","RHOXF1","PITX3","Arid3a")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_neuron_motifs_2 <- getFeatures(projNC, select = paste(neuron_motifs_2, collapse="|"), useMatrix = "MotifMatrix")

plotting_neuron_motifs_2_z <- paste0("z:",plotting_neuron_motifs_2)

motif_plot_neurons_2 <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_neuron_motifs_2_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1,
                                     randomize = F)
motif_plot_neurons_2

for (plot in seq(length(motif_plot_neurons_2))){
  plotPDF(motif_plot_neurons_2[[plot]], name = paste0("Plot-Motif-Zscores-C7-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}

# neural 2
# RFX4 PBX1 MEIS2 SOX10
# Check KROX20 genescore for these guys?

neuron_motifs_3 <- c("RFX4","PBX1","PAX3","SOX10")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_neuron_motifs_3 <- getFeatures(projNC, select = paste(neuron_motifs_3, collapse="|"), useMatrix = "MotifMatrix")

plotting_neuron_motifs_3_z <- paste0("z:",plotting_neuron_motifs_3)

motif_plot_neurons_3 <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_neuron_motifs_3_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1,
                                     randomize = F)
motif_plot_neurons_3

for (plot in seq(length(motif_plot_neurons_3))){
  plotPDF(motif_plot_neurons_3[[plot]], name = paste0("Plot-Motif-Zscores-C5-",plot),
          addDOC = FALSE, ArchRProj = projNC, width = 5, height = 5)
}

```
## F
A heatmap of the top changing motifs
```{r}
#scales::show_col(unlist(ArchRPalettes))
Cluster_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

Motif_list <- getMarkers(Cluster_Motifs, cutOff = "FDR <= 1e-50")
Motif_df <- as.data.frame(Motif_list)
top_Motifs <- Motif_df %>% dplyr::group_by(group_name) %>% dplyr::top_n(30, wt = -log10(FDR))

# Motifs by cluster mean Z-Score?
temp <- Cluster_Motifs@assays@data$Mean
# Add rownames
rownames(temp) <- Cluster_Motifs@elementMetadata$name

temp_f <- temp[unique(top_Motifs$name),]

temp_f_t <- t(temp_f)

# Who do we want to label?

label_motifs <- c("TWIST1_741", "OTX2_718","SNAI2_729",
                  "Dmbx1_366","ZEB1_412","ZBTB18_192","Pou5f1..Sox2_38",
                  "RUNX1_41","ASCL1_594","MEIS2_262","MITF_707","GRHL2_693",
                  "YY2_744","SIX1_405","TFAP2A_739","PAX6_14","NR2F2_399",
                  "RFX3_725","SP4_177","FOXD2_604","Atoh1_306","NEUROG2_162",
                  "SCRT1_727","Rarg_344","PITX2_526","SOX9_20","PHOX2B_720","ETV4_682",
                  "KLF2_496","ELK1_357","ZIC1_190","PAX3.var.2_525","PBX1_15",
                  "SOX10_407","TLX2_555","LHX5_500","LBX2_193","GBX1_372",
                  "PITX3_205","Arid3a_43","RHOXF1_210","MITF_707")

label_index <- which(colnames(temp_f_t) %in% label_motifs)
label_text <- toupper(stringr::str_split(colnames(temp_f_t)[label_index], pattern = "_", simplify = T)[,1])

heatmap1 <- ComplexHeatmap::Heatmap(matrix = temp_f_t,
                                    cluster_rows = T,
                                    cluster_columns = T,
                                    show_row_names = T,
                                    show_column_names = F,
                                    use_raster = F,
                                    col = as.character(ArchRPalettes$blueYellow),
                                    bottom_annotation = HeatmapAnnotation(
                                      names = anno_mark(at = label_index,
                                                        labels = label_text, which = "row", labels_rot = 90)))

 
plotPDF(heatmap1, name = "Plot-Motif-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 10, height = 8)

heatmap1
```

# Figure 2 - Pseudotime
## A
A two-dimensional UMAP projection colored by real timepoint-normalized pseudotime projections
```{r}
p9 <- plotTrajectory(projNC, trajectory = "hour_adj_pseudotime",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9[[1]], name = "Plot-UMAP-Pseudotime.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9[[1]]

# Pseudotime with just certain cells highlighted
# Mesenchyme

p9b <- plotTrajectory(projNC, trajectory = "C10END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9b[[1]], name = "Plot-UMAP-Pseudotime-C10END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9b[[1]]

p9c <- plotTrajectory(projNC, trajectory = "C5END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9c[[1]], name = "Plot-UMAP-Pseudotime-C5END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9c[[1]]

p9d <- plotTrajectory(projNC, trajectory = "C2END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9d[[1]], name = "Plot-UMAP-Pseudotime-C2END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9d[[1]]

p9e <- plotTrajectory(projNC, trajectory = "C9END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9e[[1]], name = "Plot-UMAP-Pseudotime-C9END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9e[[1]]

p9f <- plotTrajectory(projNC, trajectory = "C1END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9f[[1]], name = "Plot-UMAP-Pseudotime-C1END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9f[[1]]

p9g <- plotTrajectory(projNC, trajectory = "C7END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9g[[1]], name = "Plot-UMAP-Pseudotime-C7END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9g[[1]]

p9h <- plotTrajectory(projNC, trajectory = "C6END",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p9h[[1]], name = "Plot-UMAP-Pseudotime-C6END.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p9h[[1]]


# What is the average age per cluster? Does this match expectations?
avg_age = data.frame(Cluster = c("C1","C2","C3","C4","C5","C6","C7","C8","C9","C10"),
              Avg_Pseudotime = c(mean(projNC[projNC$Cell_Identity == "C1_Melanocytes"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C2_Sensory_Glia"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C3_Int_Mesenchyme"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C4_Multipotent_NC"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C5_Hindbrain_Neurons"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C6_Somatosensory_Neurons"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C7_Midbrain_Neurons"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C8_Early_Mesenchyme"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C9_Sensory_Neurons"]$hour_adj_pseudotime),
                                 mean(projNC[projNC$Cell_Identity == "C10_Late_Mesenchyme"]$hour_adj_pseudotime)))
# How many cells for each lineage?
summary(as.factor(projNC$Cell_Identity))
```

## B
Motifs that change over pseudotime.
Highlighting the mesencyhme, midbrain, somatosensory, and hindbrain neurons, sensory ganglia, and sensory neuron lineages.
First we have to isolate the cells of that lineage to perform the analysis on (differentially enriched).
Then, we aggregate transcription factor signal by family to reduce noise.
Finally, we plot a heatmap and label specific transcription factors of interest in each lineage.

```{r Family Aggregation}
matrix_agg_scores <- function(input = sm_mat, transpose = T) {
  ## A function to aggregate motif scores by family.
  
  if (transpose){
    input <- t(input)
  }
  # Convert to DF
  input <- as.data.frame(input)
  # Clean up names
  colnames(input) <- toupper(colnames(input))
  
  # GRHL
  if (any(c("GRHL1","GRHL2","GRHL3") %in% colnames(input))){
    print("Collapsing GRHL's")
    # Which are present?
    select <- c("GRHL1","GRHL2","GRHL3")[c("GRHL1","GRHL2","GRHL3") %in% colnames(input)]
    GRHL_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$GRHL_Family <- GRHL_Family
  }
  # TEAD
  if (any(c("TEAD1","TEAD2","TEAD3","TEAD4") %in% colnames(input))){
    print("Collapsing TEAD's")
    # Which are present?
    select <- c("TEAD1","TEAD2","TEAD3","TEAD4")[c("TEAD1","TEAD2","TEAD3","TEAD4") %in% colnames(input)]
    TEAD_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$TEAD_Family <- TEAD_Family
  }
  
  # TFAP2
  if (any(grep(pattern = "TFAP2", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing TAFP2's")
    select <- grep(pattern = "TFAP2", colnames(input), value = T)[grep(pattern = "TFAP2", colnames(input), value = T) %in% colnames(input)]
    TFAP2_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$TFAP2_Family <- TFAP2_Family
  }  
  
  # SANI
  if (any(c("SNAI1", "SNAI2","SNAI3") %in% colnames(input))){
    print("Collapsing SANI's")
    # Which are present?
    select <- c("SNAI1", "SNAI2","SNAI3")[c("SNAI1", "SNAI2","SNAI3") %in% colnames(input)]
    SANI_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SANI_Family <- SANI_Family
  }   
  
  # SOXB
  if (any(c("SOX1", "SOX2","SOX3") %in% colnames(input))){
    print("Collapsing SOXB's")
    # Which are present?
    select <- c("SOX1", "SOX2","SOX3")[c("SOX1", "SOX2","SOX3") %in% colnames(input)]
    SOXB_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXB_Family <- SOXB_Family
  }
  
  # SOXD
  if (any(c("SOX5", "SOX6","SOX13") %in% colnames(input))){
    print("Collapsing SOXD's")
    # Which are present?
    select <- c("SOX5", "SOX6","SOX13")[c("SOX5", "SOX6","SOX13") %in% colnames(input)]
    SOXD_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXD_Family <- SOXD_Family
  }
  
  # SOXE
  if (any(c("SOX8", "SOX9","SOX10") %in% colnames(input))){
    print("Collapsing SOXE's")
    # Which are present?
    select <- c("SOX8", "SOX9","SOX10")[c("SOX8", "SOX9","SOX10") %in% colnames(input)]
    SOXE_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXE_Family <- SOXE_Family
  }
  
  # SOXF
  if (any(c("SOX7", "SOX17","SOX18") %in% colnames(input))){
    print("Collapsing SOXF's")
    # Which are present?
    select <- c("SOX7", "SOX17","SOX18")[c("SOX7", "SOX17","SOX18") %in% colnames(input)]
    SOXF_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SOXF_Family <- SOXF_Family
  }
  
  # NR2F
  if (any(grep(pattern = "NR2F", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing NR2F's")
    select <- grep(pattern = "NR2F", colnames(input), value = T)[grep(pattern = "NR2F", colnames(input), value = T) %in% colnames(input)]
    NR2F_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$NR2F_Family <- NR2F_Family
  } 
  
  # SP
  if (any(c("SP1", "SP2","SP3","SP4","SP8","SP9") %in% colnames(input))){
    print("Collapsing SP's")
    # Which are present?
    select <- c("SP1", "SP2","SP3","SP4","SP8","SP9")[c("SP1", "SP2","SP3","SP4","SP8","SP9") %in% colnames(input)]
    SP_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SP_Family <- SP_Family
  }
  
  # FOX
  if (any(grep(pattern = "FOX", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing FOX's")
    select <- grep(pattern = "FOX", colnames(input), value = T)[grep(pattern = "FOX", colnames(input), value = T) %in% colnames(input)]
    FOX_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$FOX_Family <- FOX_Family
  } 
  
  # KLF
  if (any(grep(pattern = "KLF", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing KLF's")
    select <- grep(pattern = "KLF", colnames(input), value = T)[grep(pattern = "KLF", colnames(input), value = T) %in% colnames(input)]
    KLF_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$KLF_Family <- KLF_Family
  } 
  
  # RFX
  if (any(grep(pattern = "RFX", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing RFX's")
    select <- grep(pattern = "RFX", colnames(input), value = T)[grep(pattern = "RFX", colnames(input), value = T) %in% colnames(input)]
    RFX_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$RFX_Family <- RFX_Family
  } 
  
  # RXR
  if (any(grep(pattern = "RFR", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing RFR's")
    select <- grep(pattern = "RFR", colnames(input), value = T)[grep(pattern = "RFR", colnames(input), value = T) %in% colnames(input)]
    RFR_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$RFR_Family <- RFR_Family
  } 
  
  # SIX
  if (any(grep(pattern = "SIX", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing SIX's")
    select <- grep(pattern = "SIX", colnames(input), value = T)[grep(pattern = "SIX", colnames(input), value = T) %in% colnames(input)]
    SIX_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$SIX_Family <- SIX_Family
  }
  
  # ETV
  if (any(grep(pattern = "ETV", colnames(input), value = T) %in% colnames(input))){
    print("Collapsing ETV's")
    select <- grep(pattern = "ETV", colnames(input), value = T)[grep(pattern = "ETV", colnames(input), value = T) %in% colnames(input)]
    ETV_Family <- Matrix::rowMeans(input[,select])
    input <- input[!colnames(input) %in% select]
    input$ETV_Family <- ETV_Family
  }
  
  return(input)
}

```

```{r Mesenchyme TF over pseudotime}
# Motif scores per cell
motif_mat <- getMatrixFromProject(projNC,
                                  useMatrix = "MotifMatrix",
                                  useSeqnames = "z")
motif_mat_raw <- assay(motif_mat)
motif_mat_impute <- imputeMatrix(mat = motif_mat_raw,
                                 imputeWeights = getImputeWeights(projNC))


# Cell names along mesenchyme trajectory
mes_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C3","C8","C10")]

which(projNC$cellNames %in% mes_cells)

# Order columns by pseudotime.
mes_cells_df <- data.frame(Name = mes_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in% c("C4","C3","C8","C10")])
mes_cells_df <- mes_cells_df[order(mes_cells_df$Pseudotime),]

mes_motif_mat_impute_ord <- as.matrix(motif_mat_impute[,mes_cells_df$Name])
# Collapse by motif family.
rownames(mes_motif_mat_impute_ord) <- stringr::str_split(rownames(mes_motif_mat_impute_ord), pattern = "_", simplify = T)[,1]

mes_motif_mat_impute_ord <- t(matrix_agg_scores(mes_motif_mat_impute_ord))

# Filter to only enriched TF's
mes_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C3","C8","C10"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

mes_motif_df <- as.data.frame(getMarkers(mes_Motifs, cutOff = "FDR <= 1e-20 & MeanDiff > 2"))

selected_motifs <- toupper(unique(mes_motif_df$name))
selected_motifs <- stringr::str_split(selected_motifs, pattern = "_", simplify = T)[,1]

# Remove motifs NOT in the family-collapsed matrix.
selected_motifs <- selected_motifs[selected_motifs %in% rownames(mes_motif_mat_impute_ord)]

# Add in the family names
selected_motifs <- c(selected_motifs, grep("_Family", rownames(mes_motif_mat_impute_ord), value = T))

labels <- c("ETS1","RXRG","NR2F_Family","TWIST1",
            "ZBTB18","TFAP2_Family","JUNB",
            "FOSL1..JUN","FLI1","SOXE_Family")

labels[!labels %in% selected_motifs]

label_index <- which(selected_motifs %in% labels)


heatmap3 <- ComplexHeatmap::Heatmap(
  matrix = mes_motif_mat_impute_ord[selected_motifs,],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Mesenchyme",
  col = as.character(ArchRPalettes$blueYellow)) +
  rowAnnotation(names = anno_mark(at = label_index, labels = labels))

selected_motifs[row_order(heatmap3)]

plotPDF(heatmap3, name = "Plot-Mesenchyme-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap3
```

```{r Hindbrain Neurons TF over pseudotime}
# Cell names along Neural 1 trajectory
neural1_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C5")]

# Order columns by pseudotime.
neural1_cells_df <- data.frame(Name = neural1_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C5")])

neural1_cells_df <- neural1_cells_df[order(neural1_cells_df$Pseudotime),]

ordered_neural1 <- as.matrix(motif_mat_impute[,neural1_cells_df$Name])
# Collapse by motif family.
rownames(ordered_neural1) <- stringr::str_split(rownames(ordered_neural1), pattern = "_", simplify = T)[,1]

ordered_neural1 <- t(matrix_agg_scores(ordered_neural1))

# Filter to only enriched TF's
neural1_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C4","C5"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

neural1_motif_df <- as.data.frame(getMarkers(neural1_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1"))

selected_motifs <- toupper(unique(neural1_motif_df$name))
selected_motifs <- stringr::str_split(selected_motifs, pattern = "_", simplify = T)[,1]

# Remove motifs NOT in the family-collapsed matrix.
selected_motifs <- selected_motifs[selected_motifs %in% rownames(ordered_neural1)]

# Add in the family names
selected_motifs <- c(selected_motifs, grep("_Family", rownames(ordered_neural1), value = T))

# Choose which transcription factors to highlight
labels <- c("POU5F1..SOX2","RFX_Family","MEIS1.VAR.2","SOXB_Family",
            "PBX1","MEIS2","ZIC_Family","PBX2","MEIS3","PAX3","SOX10",
            "POU3F2","RFX3")

# Ensure labels are in selected motifs
labels[!labels %in% selected_motifs]
label_index <- which(selected_motifs %in% labels)
label_text <- toupper(stringr::str_split(rownames(ordered_neural1[selected_motifs,])[label_index], pattern = "_", simplify = T)[,1])

heatmap4 <- ComplexHeatmap::Heatmap(
  matrix = ordered_neural1[selected_motifs,],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Hindbrain Neurons",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(neural1_motif_df$name)[row_order(heatmap4)]


plotPDF(heatmap4, name = "Plot-Hindbrain-Neurons-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap4
```

```{r Midbrain Neurons TF over pseudotime}
# Cell names along Neural 2 trajectory
neural2_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C7")]

which(projNC$cellNames %in% neural2_cells)

# Order columns by pseudotime.
neural2_cells_df <- data.frame(Name = neural2_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C7")])

neural2_cells_df <- neural2_cells_df[order(neural2_cells_df$Pseudotime),]

ordered_neural2 <- as.matrix(motif_mat_impute[,neural2_cells_df$Name])

# Filter to only enriched TF's
neural2_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C4","C7"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

neural2_motif_df <- as.data.frame(getMarkers(neural2_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("OTX2_718","Pou5f1..Sox2_38","Dmbx1_366",
            "PITX2_526","SOX2_730","Arid3a_43","PHOX2A_204",
            "Shox2_211","GBX1_372","GATA1_690",
            "PITX3_205","RHOXF1_210")

label_index <- which(unique(neural2_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(neural2_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap5 <- ComplexHeatmap::Heatmap(
  matrix = ordered_neural2[unique(neural2_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Midbrain Neurons",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(neural2_motif_df$name)[row_order(heatmap5)]


plotPDF(heatmap5, name = "Plot-Midbrain-Neurons-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap5
```

```{r Sensory glia TF over pseudotime}
# Cell names along sensory glia trajectory
sensory_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C2")]

which(projNC$cellNames %in% sensory_cells)

# Order columns by pseudotime.
sensory_cells_df <- data.frame(Name = sensory_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C2")])

sensory_cells_df <- sensory_cells_df[order(sensory_cells_df$Pseudotime),]

ordered_sensory <- as.matrix(motif_mat_impute[,sensory_cells_df$Name])

# Filter to only enriched TF's
sensory_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C2"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

sensory_motif_df <- as.data.frame(getMarkers(sensory_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("GRHL2_693","ASCL1.var.2_639","ZEB1_412",
            "SNAI2_729","SIX1_405","DLX5_464",
            "MEOX2_196","ISL2_388","TEAD4_738",
            "GATA2_390")

label_index <- which(unique(sensory_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(sensory_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap6 <- ComplexHeatmap::Heatmap(
  matrix = ordered_sensory[unique(sensory_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Sensory Glia",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(sensory_motif_df$name)[row_order(heatmap6)]


plotPDF(heatmap6, name = "Plot-Sensory-Glia-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap6
```

```{r Sensory neurons TF over pseudotime}
# Cell names along sensory neuron trajectory
sensory_neuron_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C9")]

which(projNC$cellNames %in% sensory_neuron_cells)

# Order columns by pseudotime.
sensory_neuron_cells_df <- data.frame(Name = sensory_neuron_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C9")])

sensory_neuron_cells_df <- sensory_neuron_cells_df[order(sensory_neuron_cells_df$Pseudotime),]

ordered_sensory_neuron <- as.matrix(motif_mat_impute[,sensory_neuron_cells_df$Name])

# Filter to only enriched TF's
sensory_neuron_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C9"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

sensory_neuron_motif_df <- as.data.frame(getMarkers(sensory_neuron_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("Foxd3_8","FOXE1_472","FOXO4_334","HAND2_646",
            "NEUROG2.var.2_650","ATOH1.var.2_455","NEUROD1_395",
            "Ascl2_305","GATA5_605")

label_index <- which(unique(sensory_neuron_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(sensory_neuron_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap7 <- ComplexHeatmap::Heatmap(
  matrix = ordered_sensory_neuron[unique(sensory_neuron_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Sensory Neurons",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(sensory_neuron_motif_df$name)[row_order(heatmap7)]

plotPDF(heatmap7, name = "Plot-Sensory-neuron-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap7
```

```{r Somatosensory Neurons TF over pseudotime}
# Cell names along Neural 2 trajectory
neural3_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C6")]

which(projNC$cellNames %in% neural3_cells)

# Order columns by pseudotime.
neural3_cells_df <- data.frame(Name = neural3_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C6")])

neural3_cells_df <- neural3_cells_df[order(neural3_cells_df$Pseudotime),]

ordered_neural3 <- as.matrix(motif_mat_impute[,neural3_cells_df$Name])

# Filter to only enriched TF's
neural3_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C4","C6"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

neural3_motif_df <- as.data.frame(getMarkers(neural3_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("OTX2_718","Pou5f1..Sox2_38","Dmbx1_366",
            "PITX2_526","SOX2_730","Arid3a_43","PHOX2A_204",
            "Shox2_211","GBX1_372","GATA1_690", "GSC_139",
            "PITX3_205","RHOXF1_210","LBX2_193","EN1_128",
            "LHX5_500","TLX2_555","Sox3_72","Msx3_199", "Alx4_338")

label_index <- which(unique(neural3_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(neural3_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap8 <- ComplexHeatmap::Heatmap(
  matrix = ordered_neural3[unique(neural3_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Somatosensory Neurons",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(neural3_motif_df$name)[row_order(heatmap8)]


plotPDF(heatmap8, name = "Plot-Somatosensory-Neurons-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap8
```

```{r Melanocytes TF over pseudotime}
# Cell names along Neural 2 trajectory
melano_cells <- projNC$cellNames[projNC$Clusters %in% c("C4","C1")]

which(projNC$cellNames %in% melano_cells)

# Order columns by pseudotime.
melano_cells_df <- data.frame(Name = melano_cells,
                           Pseudotime = projNC$hour_adj_pseudotime[projNC$Clusters %in%  c("C4","C1")])

melano_cells_df <- melano_cells_df[order(melano_cells_df$Pseudotime),]

ordered_melano <- as.matrix(motif_mat_impute[,melano_cells_df$Name])

# Filter to only enriched TF's
melano_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    useGroups = c("C4","C1"),
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

melano_motif_df <- as.data.frame(getMarkers(melano_Motifs, cutOff = "FDR <= 1e-15 & MeanDiff > 1.5"))

labels <- c("KLF14_235","SP1_632","NFYA_397","ELK3_248",
            "ETV3_253","YY2_744","FOSL1..JUN.var.2_422",
            "Creb5_326","ATF3_595","MITF_707","MLXIPL_157",
            "EGR3_228","BHLHE41_122","TCFL5_635","TCF3_734")

label_index <- which(unique(melano_motif_df$name) %in% labels)
label_text <- stringr::str_split(unique(melano_motif_df$name)[label_index], pattern = "_", simplify = T)[,1]


heatmap9 <- ComplexHeatmap::Heatmap(
  matrix = ordered_melano[unique(melano_motif_df$name),],
  cluster_rows = T, cluster_columns = F, show_column_names = F,
  show_row_names = F, use_raster = F, row_title = "chromVAR z-score",
  column_title = "Cells ordered in Pseudotime: Melanocytes",
  col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

unique(melano_motif_df$name)[row_order(heatmap9)]


plotPDF(heatmap9, name = "Plot-Melanocyte-Motif-Pseudotime-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 5)

heatmap9
```

## C
Individual line plots of just top Motifs for Mesenchyme branch
```{r Line plots for Mesenchyme over pseudotime}
interesting_mesenchyme_factors <- c("ETS1","RXRG","NR2F_Family","TWIST1",
            "ZBTB18","TFAP2_Family","JUNB",
            "FOSL1..JUN","FLI1","SOXE_Family")


# Motif Matrix for a gene within a particular trajectory
# TWIST1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","TWIST1_741"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-TWIST1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# NR2F1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","NR2F1_220"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-NR2F1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# ZBTB18
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","ZBTB18_192"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-ZBTB18-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# ETS1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","ETS1_251"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-ETS1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# FLI1
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","FLI1_133"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-FLI1-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# TFAP2B
p10 <- plotTrajectory(projNC, trajectory = "C10END", colorBy = "MotifMatrix",
                      name = paste0("z:","TFAP2B_300"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(p10[[2]], name = "Plot-TFAP2B-MOTIF-Mesencyme-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

```
```{r Line plots for Somatosensory Neurons over pseudotime}
interesting_somato_factors <- c("PITX3","OTX2","LBX2","GBX1","TLX2","LHX5",
                                "EN1","SOX3","PHOX2A")

# Motif Matrix for a gene within a particular trajectory
# PHOX2A
s10 <- plotTrajectory(projNC, trajectory = "C6END", colorBy = "MotifMatrix",
                      name = paste0("z:","PHOX2A_204"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(s10[[2]], name = "Plot-PHOX2A-MOTIF-Somatosensory-Neurons-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# SOX3
s10 <- plotTrajectory(projNC, trajectory = "C6END", colorBy = "MotifMatrix",
                      name = paste0("z:","Sox3_72"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(s10[[2]], name = "Plot-SOX3-MOTIF-Somatosensory-Neurons-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# LBX2
s10 <- plotTrajectory(projNC, trajectory = "C6END", colorBy = "MotifMatrix",
                      name = paste0("z:","LBX2_193"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(s10[[2]], name = "Plot-LBX2-MOTIF-Somatosensory-Neurons-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# GBX1
s10 <- plotTrajectory(projNC, trajectory = "C6END", colorBy = "MotifMatrix",
                      name = paste0("z:","GBX1_372"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(s10[[2]], name = "Plot-GBX1-MOTIF-Somatosensory-Neurons-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# TLX2
s10 <- plotTrajectory(projNC, trajectory = "C6END", colorBy = "MotifMatrix",
                      name = paste0("z:","TLX2_555"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(s10[[2]], name = "Plot-TLX2-MOTIF-Somatosensory-Neurons-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)

# LHX5
s10 <- plotTrajectory(projNC, trajectory = "C6END", colorBy = "MotifMatrix",
                      name = paste0("z:","LHX5_500"), continuousSet = "blueYellow",
                      plotAs = "points", addArrow = F)
plotPDF(s10[[2]], name = "Plot-LHX5-MOTIF-Somatosensory-Neurons-Pseudotime-Line.pdf",
        ArchRProj = projNC, addDOC = FALSE, width = 6, height = 4)


```

## D
Individual line plots of top genes in Motif and Genescore.



# Figure 3
Genescores

## A
Investigation of Genescores.

TODO - Genescores using this new algorithm seems worse than previous implemention from Cicero. Perhaps I should compare Cicero vs ArchR Coaccessibility
1.  It seems that ArchR gives 'cleaner' CoAccessibility.
```{r}
projNC_noC1 <- projNC[!projNC$Clusters == "C1"]

# Genescores on non-imputed data.
markersGS <- getMarkerFeatures(
    ArchRProj = projNC_noC1, 
    useMatrix = "GeneScoreMatrix", 
    groupBy = "Clusters",
    bias = c("TSSEnrichment", "log10(nFrags)"),
    #normBy = "ReadsInPeaks",
    testMethod = "ttest")

strong_GS_markers <- as.data.frame(getMarkers(markersGS, cutOff = "FDR <= 0.1 & Log2FC >= 1"))

# Remove C1 overweight
marker_list <- getMarkers(markersGS, cutOff = "FDR <= 0.1 & Log2FC >= 1")
C1_markers <- paste0(marker_list$C1$seqnames,":", marker_list$C1$name)

C1_markers_logical <- ifelse(stringr::str_split(stringr::str_split(C1_markers, pattern = ":", simplify = T)[,2], pattern = "_", simplify = T)[,1] == "NA",
       FALSE, TRUE)

C1_markers_filtered <- C1_markers[C1_markers_logical]

# Focus Labels
GSMarkerLabel <- c("TWIST1","PRRX2","SOX10","COL9A3","VEGFD","TMEM120A","SLC6A7","PRRX1","CRABP1",
                   "S100A6","S100A4","CLDN9","DLX4","HOXB2","HOXB3","FAM181A","RAX","WNT3","WNT8B",
                   "AQP2","GNRH1","OR14J1L57","CER1")
label_index <- which(unique(strong_GS_markers$name) %in% GSMarkerLabel)
label_text <- stringr::str_split(unique(strong_GS_markers$name)[label_index], pattern = "_", simplify = T)[,1]

# Heatmap of all differentially enriched genes.
genescore_mat <- getMatrixFromProject(projNC, useMatrix = "GeneScoreMatrix")
genescore_mat@assays@data$GeneScoreMatrix@Dimnames[[1]] <- genescore_mat@elementMetadata$name

genescore_mat_f <- as.matrix(genescore_mat@assays@data$GeneScoreMatrix)

# Order columns by pseudotime.
genescore_cells_df <- data.frame(Name = projNC$cellNames,
                           Pseudotime = projNC$hour_adj_pseudotime)

genescore_cells_df <- genescore_cells_df[order(genescore_cells_df$Pseudotime),]

ordered_genescore <- as.matrix(genescore_mat_f[,genescore_cells_df$Name])

rownames(ordered_genescore) %in% strong_GS_markers$name


# heatmap10 <- ComplexHeatmap::Heatmap(
#   matrix = ordered_genescore[1:100,1:100],
#   cluster_rows = T, cluster_columns = F,
#   show_column_names = F, show_row_names = F,
#   use_raster = T, row_title = "chromVAR z-score",
#   column_title = "Top GeneScores",
#   col = as.character(ArchRPalettes$blueYellow)) + 
#   rowAnnotation(names = anno_mark(at = label_index, labels = label_text))
# 
# plotPDF(heatmap10, name = "GeneScore_Heatmap_Labeled.pdf",ArchRProj = projNC, width = 8, height = 6, addDOC = F)
# 
# heatmap10

plotMarkerHeatmap(seMarker = markersGS, cutOff = "FDR <= 0.1 & Log2FC >= 1",
                  nLabel = 5, plotLog2FC = T, grepExclude = "NA", transpose = T,
                  pal = ArchRPalettes$horizonExtra)

plotPDF(genescore_heatmap, name = "GeneScore_Heatmap_Labeled.pdf", ArchRProj = projNC, width = 8, height = 6, addDOC = F)
```

## B
DiffTF-like correlation of TF expression to accessibility of peaks that contain their motif.

We can use this to identify transcription factors of interest in each lineage.
Interested in FOXD3 'biomodal' activity.

```{r Across the entire dataset}
seGroupMotif <- getGroupSE(ArchRProj = projNC, useMatrix = "MotifMatrix", groupBy = "Clusters")
seZ <- seGroupMotif[rowData(seGroupMotif)$seqnames=="z",]

rowData(seZ)$maxDelta <- lapply(seq_len(ncol(seZ)), function(x){
  rowMaxs(assay(seZ) - assay(seZ)[,x])
}) %>% Reduce("cbind", .) %>% rowMaxs

corGSM_MM <- correlateMatrices(
    ArchRProj = projNC,
    useMatrix1 = "GeneScoreMatrix",
    useMatrix2 = "MotifMatrix",
    reducedDims = "IterativeLSI")

corGSM_MM$maxDelta <- rowData(seZ)[match(corGSM_MM$MotifMatrix_name, rowData(seZ)$name), "maxDelta"]

corGSM_MM <- corGSM_MM[order(abs(corGSM_MM$cor), decreasing = TRUE), ]
corGSM_MM <- corGSM_MM[which(!duplicated(gsub("\\-.*","",corGSM_MM[,"MotifMatrix_name"]))), ]
corGSM_MM$TFRegulator <- "NO"

# Positive
corGSM_MM$TFRegulator[which(corGSM_MM$cor > 0.3 & corGSM_MM$padj < 0.1 & corGSM_MM$maxDelta > quantile(corGSM_MM$maxDelta, 0.7))] <- "Positive"
# Negative
corGSM_MM$TFRegulator[which(corGSM_MM$cor < -0.3 & corGSM_MM$padj < 0.1 & corGSM_MM$maxDelta > quantile(corGSM_MM$maxDelta, 0.7))] <- "Negative"

sort(corGSM_MM[corGSM_MM$TFRegulator=="Positive",1])

corGSM_MM$label <- ifelse(corGSM_MM$TFRegulator %in% c("Positive","Negative"), corGSM_MM$GeneScoreMatrix_name, "")

p13 <- ggplot(data.frame(corGSM_MM), aes(cor, maxDelta, color = TFRegulator, label = label)) +
  geom_point() + 
  theme_ArchR() +
  geom_vline(xintercept = 0, lty = "dashed") + 
  scale_color_manual(values = c("NO"="darkgrey", "Positive"="firebrick3", "Negative"="blue")) +
  xlab("Correlation To Gene Score") +
  ylab("Max TF Motif Delta") +
  scale_y_continuous(
    expand = c(0,0), 
    limits = c(0, max(corGSM_MM$maxDelta)*1.05)
  ) + geom_label_repel(max.overlaps = 20, box.padding = 0.25)

plotPDF(p13, name = "Plot-DiffTFlike-GeneScore-TF-Correlation.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 7)

p13

write_csv(as.data.frame(corGSM_MM), "Figure_Exports/DiffTFlike-GeneScore-TF-Correlation-Table.csv")
```

```{r Mesencyhme vs Sensory Neurons}
# Mesenchyme
projNC_mes <- projNC[projNC$Cell_Identity %in% 
                       c("C4_Multipotent_NC","C8_Early_Mesenchyme",
                         "C3_Int_Mesenchyme","C10_Late_Mesenchyme"),]
  

seGroupMotif_mes <- getGroupSE(ArchRProj = projNC_mes, useMatrix = "MotifMatrix", groupBy = "Clusters")
seZ_mes <- seGroupMotif_mes[rowData(seGroupMotif_mes)$seqnames=="z",]

rowData(seZ_mes)$maxDelta <- lapply(seq_len(ncol(seZ_mes)), function(x){
  rowMaxs(assay(seZ_mes) - assay(seZ_mes)[,x])
}) %>% Reduce("cbind", .) %>% rowMaxs

corGSM_MM_mes <- correlateMatrices(
    ArchRProj = projNC_mes,
    useMatrix1 = "GeneScoreMatrix",
    useMatrix2 = "MotifMatrix",
    reducedDims = "IterativeLSI")

corGSM_MM_mes$maxDelta <- rowData(seZ_mes)[match(corGSM_MM_mes$MotifMatrix_name, rowData(seZ_mes)$name), "maxDelta"]

corGSM_MM_mes <- corGSM_MM_mes[order(abs(corGSM_MM_mes$cor), decreasing = TRUE), ]
corGSM_MM_mes <- corGSM_MM_mes[which(!duplicated(gsub("\\-.*","",corGSM_MM_mes[,"MotifMatrix_name"]))), ]
corGSM_MM_mes$TFRegulator <- "NO"

# Positive
corGSM_MM_mes$TFRegulator[which(corGSM_MM_mes$cor > 0.3 & corGSM_MM_mes$padj < 0.1 & corGSM_MM_mes$maxDelta > quantile(corGSM_MM_mes$maxDelta, 0.7))] <- "Positive"
# Negative
corGSM_MM_mes$TFRegulator[which(corGSM_MM_mes$cor < -0.3 & corGSM_MM_mes$padj < 0.1 & corGSM_MM_mes$maxDelta > quantile(corGSM_MM_mes$maxDelta, 0.7))] <- "Negative"

sort(corGSM_MM_mes[corGSM_MM_mes$TFRegulator=="Positive",1])

corGSM_MM_mes$label <- ifelse(corGSM_MM_mes$TFRegulator %in% c("Positive","Negative"), corGSM_MM_mes$GeneScoreMatrix_name, "")

p14 <- ggplot(data.frame(corGSM_MM_mes), aes(cor, maxDelta, color = TFRegulator, label = label)) +
  geom_point() + 
  theme_ArchR() +
  geom_vline(xintercept = 0, lty = "dashed") + 
  scale_color_manual(values = c("NO"="darkgrey", "Positive"="firebrick3", "Negative"="blue")) +
  xlab("Correlation To Gene Score") +
  ylab("Max TF Motif Delta") +
  scale_y_continuous(
    expand = c(0,0), 
    limits = c(0, max(corGSM_MM_mes$maxDelta)*1.05)
  ) + geom_label_repel(max.overlaps = 20, box.padding = 0.25)

plotPDF(p14, name = "Plot-Mesencyme-GeneScore-TF-Correlation.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 7)

p14

projNC_senseNeuron <- projNC[projNC$Cell_Identity %in% 
                       c("C4_Multipotent_NC","C9_Sensory_Neurons"),]
  

seGroupMotif_senseNeuron <- getGroupSE(ArchRProj = projNC_senseNeuron, useMatrix = "MotifMatrix", groupBy = "Clusters")
seZ_senseNeuron <- seGroupMotif_senseNeuron[rowData(seGroupMotif_senseNeuron)$seqnames=="z",]

rowData(seZ_senseNeuron)$maxDelta <- lapply(seq_len(ncol(seZ_senseNeuron)), function(x){
  rowMaxs(assay(seZ_senseNeuron) - assay(seZ_senseNeuron)[,x])
}) %>% Reduce("cbind", .) %>% rowMaxs

corGSM_MM_senseNeuron <- correlateMatrices(
    ArchRProj = projNC_senseNeuron,
    useMatrix1 = "GeneScoreMatrix",
    useMatrix2 = "MotifMatrix",
    reducedDims = "IterativeLSI")

corGSM_MM_senseNeuron$maxDelta <- rowData(seZ_senseNeuron)[match(corGSM_MM_senseNeuron$MotifMatrix_name, rowData(seZ_senseNeuron)$name), "maxDelta"]

corGSM_MM_senseNeuron <- corGSM_MM_senseNeuron[order(abs(corGSM_MM_senseNeuron$cor), decreasing = TRUE), ]
corGSM_MM_senseNeuron <- corGSM_MM_senseNeuron[which(!duplicated(gsub("\\-.*","",corGSM_MM_senseNeuron[,"MotifMatrix_name"]))), ]
corGSM_MM_senseNeuron$TFRegulator <- "NO"

# Positive
corGSM_MM_senseNeuron$TFRegulator[which(corGSM_MM_senseNeuron$cor > 0.3 & corGSM_MM_senseNeuron$padj < 0.1 & corGSM_MM_senseNeuron$maxDelta > quantile(corGSM_MM_senseNeuron$maxDelta, 0.7))] <- "Positive"
# Negative
corGSM_MM_senseNeuron$TFRegulator[which(corGSM_MM_senseNeuron$cor < -0.3 & corGSM_MM_senseNeuron$padj < 0.1 & corGSM_MM_senseNeuron$maxDelta > quantile(corGSM_MM_senseNeuron$maxDelta, 0.7))] <- "Negative"

sort(corGSM_MM_senseNeuron[corGSM_MM_senseNeuron$TFRegulator=="Positive",1])

corGSM_MM_senseNeuron$label <- ifelse(corGSM_MM_senseNeuron$TFRegulator %in% c("Positive","Negative"), corGSM_MM_senseNeuron$GeneScoreMatrix_name, "")

p15 <- ggplot(data.frame(corGSM_MM_senseNeuron), aes(cor, maxDelta, color = TFRegulator, label = label)) +
  geom_point() + 
  theme_ArchR() +
  geom_vline(xintercept = 0, lty = "dashed") + 
  scale_color_manual(values = c("NO"="darkgrey", "Positive"="firebrick3", "Negative"="blue")) +
  xlab("Correlation To Gene Score") +
  ylab("Max TF Motif Delta") +
  scale_y_continuous(
    expand = c(0,0), 
    limits = c(0, max(corGSM_MM_senseNeuron$maxDelta)*1.05)
  ) + geom_label_repel(max.overlaps = 20, box.padding = 0.25)

plotPDF(p15, name = "Plot-SensoryNeuron-GeneScore-TF-Correlation.pdf", ArchRProj = projNC, addDOC = FALSE, width = 6, height = 7)

p15

# What are the largest differences between shared transcription factors?
# First, see how many are classified differently.

mes_summary <- as.data.frame(corGSM_MM_mes[corGSM_MM_mes$TFRegulator %in% c("Positive","Negative"), c("cor","label")])

sn_summary <- as.data.frame(corGSM_MM_senseNeuron[corGSM_MM_senseNeuron$TFRegulator %in% c("Positive","Negative"), c("cor","label")])

correlation_comparison <- merge(mes_summary, sn_summary, by = "label", all.x = TRUE, all.y = TRUE)
```

## E 
Integrative pseudo-time analyses.
Let's find the best correlated Gene to Motif scores in an unbaised way.
```{r Integration of Genescore and Motifs in pseudotime}
# For each lineage, find the best correlated TF+Genes

for(lineage in c("C1END","C2END","C5END","C6END","C7END","C9END","C10END")){
  END_MM  <- getTrajectory(ArchRProj = projNC, name = lineage, useMatrix = "MotifMatrix", log2Norm = FALSE)
  END_GSM <- getTrajectory(ArchRProj = projNC, name = lineage, useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
  END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.4)

  END_trajGSM2 <- END_GSM[END_MM_corGSM_MM[[1]]$name2, ]
  END_trajMM2 <- END_MM[END_MM_corGSM_MM[[1]]$name1, ]
  trajCombined <- END_trajGSM2
  assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
  combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
  rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
  ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
  ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
  combined_cor_heatmap <- ht1 + ht2
  plotPDF(combined_cor_heatmap, name = paste0(lineage,"_TF-Gene_Correlation.pdf"), ArchRProj = projNC,
           addDOC = FALSE)
}

# This was a great initial pass. After a bunch of literature review, let's focus on the highlighted factors.

# Mesenchyme focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C10END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C10END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.6, varCutOff2 = 0.6, maxDist = 500000)

# Filter for TF's of interest 
mes_tfs <- c("z:ZBTB18_192","z:TWIST1_741","z:TFAP2B_300","z:NR2F1_220","z:ETS1_251","z:FLI1_133")
END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% mes_tfs,]

END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))


combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)

rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)

combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Mesenchyme_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)

# Sensory Glia Focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C2END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C2END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.6, varCutOff2 = 0.6, maxDist = 500000)

# Filter for TF's of interest 
sensory_glia_factors <- c("z:SIX1_405","z:DLX5_464","z:ZEB1_412","z:SNAI2_729","z:GATA2_390","z:ZIC5_561")

END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% sensory_glia_factors,]
END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Sensory_Glia_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)

# Sensory Neuron Focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C9END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C9END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.2, varCutOff2 = 0.2, maxDist = 500000)

# Filter for TF's of interest 
sensory_neuron_factors <- c("z:GATA5_605","z:FOXP1_688","z:SOX10_407","z:ZIC5_561","z:SIX1_405","z:FOXC2_333")

END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% sensory_neuron_factors,]
END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Sensory_Neurons_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)

# Melanocyte Focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C1END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C1END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.2, varCutOff2 = 0.2, maxDist = 500000)

# Filter for TF's of interest 
melano_factors <- c("z:NFYA_397","z:KLF11_493","z:SP1_632","z:YY2_744","z:ETV3_253","z:ATF2_640")

END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% melano_factors,]
END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Melanocyte_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)


# Hindbrain Neuron Focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C5END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C5END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.2, varCutOff2 = 0.2, maxDist = 500000)

# Filter for TF's of interest 
hindbrain_factors <- c("z:RFX4_287","z:MEIS1_261","z:SOX2_730","z:MEIS2_262","z:RFX2_283","z:RFX3_725")

END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% hindbrain_factors,]
END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Hindbrain_Neurons_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)

# Midbrain Neuron Focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C7END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C7END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.2, varCutOff2 = 0.2, maxDist = 500000)

# Filter for TF's of interest 
midbrain_factors <- c("z:OTX2_718","z:GATA4_691","z:GLI3_474","z:RFX4_287","z:PAX7_174","z:TBX6_545")

END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% midbrain_factors,]
END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Midbrain_Neurons_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)

# Somatosensory Neuron Focus
END_MM  <- getTrajectory(ArchRProj = projNC, name = "C5END", useMatrix = "MotifMatrix", log2Norm = FALSE)
END_GSM <- getTrajectory(ArchRProj = projNC, name = "C5END", useMatrix = "GeneScoreMatrix", log2Norm = FALSE)
END_MM_corGSM_MM <- correlateTrajectories(END_MM, END_GSM, corCutOff = 0.1, varCutOff1 = 0.1, varCutOff2 = 0.1, maxDist = 500000)

# Filter for TF's of interest 
somato_factors <- c("z:RFX4_287","z:SOX2_730","z:OTX2_718","z:LHX5_500","z:MEIS1_261","z:BHLHE40_307")

END_MM_corGSM_MM_filt <- END_MM_corGSM_MM[[1]][END_MM_corGSM_MM$correlatedMappings$name1 %in% somato_factors,]
END_trajGSM2 <- END_GSM[END_MM_corGSM_MM_filt$name2, ]
END_trajMM2 <- END_MM[END_MM_corGSM_MM_filt$name1, ]
trajCombined <- END_trajGSM2
assay(trajCombined, withDimnames = F) <- t(apply(assay(END_trajGSM2), 1, scale)) + t(apply(assay(END_trajMM2), 1, scale))
combinedMat <- plotTrajectoryHeatmap(trajCombined, returnMat = TRUE, varCutOff = 0)
rowOrder <- match(rownames(combinedMat), rownames(END_trajGSM2))
ht1 <- plotTrajectoryHeatmap(END_trajGSM2,  pal = paletteContinuous(set = "horizonExtra"),  varCutOff = 0, rowOrder = rowOrder)
ht2 <- plotTrajectoryHeatmap(END_trajMM2, pal = paletteContinuous(set = "solarExtra"), varCutOff = 0, rowOrder = rowOrder)
combined_cor_heatmap <- ht1 + ht2

plotPDF(combined_cor_heatmap, name = paste0("Somatosensory_Neurons_TF-Gene_Correlation_Heatmap.pdf"), ArchRProj = projNC,
           addDOC = FALSE)

```



