---
title: "R Notebook"
output:
  html_notebook: 
    toc: yes
    code_folding: hide
    fig_width: 8
    fig_height: 6
editor_options:
  chunk_output_type: inline
---


```{r Loading libraries and data, message=FALSE, warning=FALSE}
library(ArchR)
library(knitr)
library(TFBSTools)
library(ComplexHeatmap)
library(ggplot2)

addArchRThreads(threads = 62) 
set.seed(1)
projNC <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_data_final/")
projNC@projectMetadata$outputDirectory <- "./Figure_Exports"
```

# Figure 1 - Dimensionality Reduction & Clustering

## A
A picture of a HH16/18 chicken embryo electroporated with the TFAP2aE1 enhancer alongside a schematic of single-cell ATAC-Seq.

## B
A two-dimensional UMAP projection colored by original sample identity (timepoint).
```{r}
p1 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData",
                    name = "Sample", plotAs = "points",
                    pal = ArchRPalettes$bear, size = 0.2, rastr = F)
plotPDF(p1, name = "Plot-UMAP-Sample.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p1
```

## C
A two-dimensional UMAP projection followed by clustering in ArchR.
```{r}
p2 <- plotEmbedding(projNC, embedding = "UMAP", colorBy = "cellColData", 
                    name = "Clusters", plotAs = "points", size = 0.2, rastr = F)
plotPDF(p2, name = "Plot-UMAP-Clusters.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p2
```


## D
A customized plot of enriched chromVAR motifs in the UMAP projection.
```{r}
motifs <- c("Pou5f1..Sox2","RUNX1","ASCL1","MEIS2","MITF","TWIST1","GRHL2","OTX2")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_motifs <- getFeatures(projNC, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z <- paste0("z:",plotting_motifs)

# Problem with threads > 1...
motif_plot_nocutoff <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)


combined <- cowplot::plot_grid(motif_plot_nocutoff[[1]],motif_plot_nocutoff[[2]],
             motif_plot_nocutoff[[4]],motif_plot_nocutoff[[3]],
             motif_plot_nocutoff[[8]],motif_plot_nocutoff[[7]],
             motif_plot_nocutoff[[10]],motif_plot_nocutoff[[9]], nrow = 2)

plotPDF(combined, name = "Plot-Motif-Zscores.pdf", ArchRProj = projNC, addDOC = FALSE, width = 16, height = 16)

combined
```
# Figure S1 - QC Metrics
We saved a pre-filtering version of our object, and use it here to show pre-filtering QC.

## Prep

```{r}
library(ArchR)
library(knitr)
library(TFBSTools)
library(ComplexHeatmap)
library(ggplot2)
library(BSgenome.Ggallus.ENSEMBL.galGal6)
library(org.Gg.eg.db)
library(GenomicFeatures)
library(dplyr)
library(qdapTools)

addArchRThreads(threads = 32) 
set.seed(1)

HH6_dir <- "/data/Austin/10XscATAC/HH6_10xscATAC_Lane1/outs/"
HH8_dir <- "/data/Austin/10XscATAC/HH8_10xscATAC_Lane1_Lane2/outs/"
HH10_dir <- "/data/Austin/10XscATAC/HH10_10xscATAC_Lane1_Lane2/outs/"
HH12_dir <- "/data/Austin/10XscATAC/HH12_10xscATAC_Lane1_Lane2/outs/"
HH14_dir <- "/data/Austin/10XscATAC/HH14_10xscATAC_Lane1_Lane2/outs/"
HH16_dir <- "/data/Austin/10XscATAC/HH16_10xscATAC_Lane1/outs/"
HH18_dir <- "/data/Austin/10XscATAC/HH18_10xscATAC_Lane1/outs/"
libraries <- c("HH6","HH8","HH10","HH12","HH14","HH16","HH18")

# Genome annotation
arch_gg6 <- createGenomeAnnotation(genome = BSgenome.Ggallus.ENSEMBL.galGal6, filter = F)
# Manually remove noncontigous chromosomes.
arch_gg6$chromSizes <- arch_gg6$chromSizes[1:35]
TxDb_galGal6 <- makeTxDbFromEnsembl(organism = "Gallus gallus", release = 103)
arch_gg6_genes <- createGeneAnnotation(TxDb = TxDb_galGal6, OrgDb = org.Gg.eg.db, annoStyle = "ENSEMBL")

doublet_cells <- as.character(read.csv(file = "doublet_cells.csv", row.names = 1)[,1])

projNC_pre <- loadArchRProject(path = "/data/Austin/10XscATAC/ArchR_Prefiltered/")

projNC_pre@projectMetadata$outputDirectory <- "./Figure_Exports"
projNC_pre$DoubletCall <- ifelse(projNC_pre$cellNames %in% doublet_cells,
                                 "Doublet","Singlet")
```

## A
A Fragment-size distribution plot 
```{r}
p4 <- plotFragmentSizes(ArchRProj = projNC_pre)
plotPDF(p4, name = "Prefiltered-Plot-Sample-Fragment-Dist.pdf", ArchRProj = projNC_pre,
        addDOC = FALSE, width = 8, height = 8)
p4
```

## B
A TSS enrichment plot and FRiP score plot
```{r}
p5 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "TSSEnrichment",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 3.5)

plotPDF(p5, name = "Prefiltered-Plot-Sample-TSS.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p5

p6 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "FRIP",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 0.4)

plotPDF(p6, name = "Prefiltered-Plot-Sample-FRiP.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p6
```
## C 
Doublet Detection Plot

```{r}
p7 <- plotEmbedding(ArchRProj = projNC_pre, colorBy = "cellColData", name = "DoubletCall", embedding = "UMAP", size = 1, baseSize = 14, plotAs = "points")

plotPDF(p7, name = "Prefiltered-Plot-UMAP-Doublet.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)

p7
```

## D
Depth Plot

```{r}
p8 <- plotGroups(
    ArchRProj = projNC_pre, 
    groupBy = "Sample", 
    colorBy = "cellColData", 
    name = "log10(nFrags)",
    plotAs = "violin",
    alpha = 0.4,
    addBoxPlot = TRUE,
    discreteSet = "bear"
   ) + geom_hline(linetype = 5, size = 1,show.legend = F, yintercept = 3.3)
plotPDF(p8, name = "Prefiltered-Plot-Sample-Depth.pdf", ArchRProj = projNC_pre, addDOC = FALSE, width = 8, height = 8)
p8
```
## E
Extra motif plots of markers
```{r}
motifs <- c("YY2","SIX3","TFAP2A","PAX6","FOXD3","NR2F2","RFX3")

grey_red <- ArchRPalettes$whiteRed
grey_red[1] <- "grey"

plotting_motifs <- getFeatures(projNC, select = paste(motifs, collapse="|"), useMatrix = "MotifMatrix")

plotting_motifs_z <- paste0("z:",plotting_motifs)

# Problem with threads > 1...
motif_plot_supp <- plotEmbedding(projNC,
                                     embedding = "UMAP",
                                     colorBy = "MotifMatrix",
                                     name = plotting_motifs_z,
                                     plotAs = "points",
                                     sampleCells = NULL,
                                     rastr = F,
                                     keepAxis = F,
                                     threads = 1)

combined2 <- cowplot::plot_grid(motif_plot_supp[[9]],motif_plot_supp[[3]],
             motif_plot_supp[[8]],motif_plot_supp[[2]],
             motif_plot_supp[[1]],motif_plot_supp[[6]],
             motif_plot_supp[[7]], nrow = 2)

plotPDF(combined2, name = "Plot-Motif-Zscores-supp.pdf", ArchRProj = projNC, addDOC = FALSE, width = 16, height = 16)

combined2
```
## F
A heatmap of the top changing motifs
```{r}
#scales::show_col(unlist(ArchRPalettes))
Cluster_Motifs <- getMarkerFeatures(
    ArchRProj = projNC, 
    useMatrix = "MotifMatrix", 
    groupBy = "Clusters",
    bias = c("log10(nFrags)"),
    maxCells = 750,
    testMethod = "wilcoxon",
    useSeqnames = "z", threads = 1)

Motif_list <- getMarkers(Cluster_Motifs, cutOff = "FDR <= 1e-50")
Motif_df <- as.data.frame(Motif_list)
top_Motifs <- Motif_df %>% dplyr::group_by(group_name) %>% dplyr::top_n(30, wt = -log10(FDR))

# Motifs by cluster mean Z-Score?
temp <- Cluster_Motifs@assays@data$Mean
# Add rownames
rownames(temp) <- Cluster_Motifs@elementMetadata$name

temp_f <- temp[unique(top_Motifs$name),]

# Who do we want to label?

label_motifs <- c("TWIST1_741", "OTX2_718","SNAI2_729",
                  "Dmbx1_366","ZEB1_412","ZBTB18_192","Pou5f1..Sox2_38",
                  "RUNX1_41","ASCL1_594","MEIS2_262","MITF_707","GRHL2_693",
                  "YY2_744","SIX1_405","TFAP2A_739","PAX6_14","NR2F2_399",
                  "RFX3_725","SP4_177","FOXD2_604","Atoh1_306","NEUROG2_162",
                  "SCRT1_727","Rarg_344","PITX2_526","SOX9_20","PHOX2B_720","ETV4_682",
                  "KLF2_496","ELK1_357")

label_index <- which(rownames(temp_f) %in% label_motifs)
label_text <- toupper(stringr::str_split(rownames(temp_f)[label_index], pattern = "_", simplify = T)[,1])

heatmap1 <- ComplexHeatmap::Heatmap(matrix = temp_f, cluster_rows = T, 
                        cluster_columns = T,
                        show_row_names = F,
                        use_raster = F, col = as.character(ArchRPalettes$blueYellow)) + 
  rowAnnotation(names = anno_mark(at = label_index, labels = label_text))

plotPDF(heatmap1, name = "Plot-Motif-Heatmap.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 10)

heatmap1
```

# Figure 2 - Pseudotime
## D
A two-dimensional UMAP projection colored by real timepoint-normalized pseudotime projections
```{r}
p3 <- plotTrajectory(projNC, trajectory = "hour_adj_pseudotime",
                          colorBy = "cellColData", name = "hour_adj_pseudotime",
                          plotAs = "points", addArrow = F, embedding = "UMAP", rastr = F)
plotPDF(p3[[1]], name = "Plot-UMAP-Pseudotime.pdf", ArchRProj = projNC, addDOC = FALSE, width = 8, height = 8)

p3[[1]]
```

